---
title: "exosome KRAS inhibitor analysis"
author: 'Roman E. Reggiardo'
date: '12/16/2020'
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r, setup, include=FALSE}
knitr::opts_chunk$set(fig.width = 8, collapse = T)
knitr::opts_chunk$set(message = F)
library(here)
print(here())
```

# setup:
## loading
```{r loading}
suppressPackageStartupMessages({
  library(tidyverse)
  library(here)
  library(ggsci)
  library(ggpubr)
  library(ggsignif)
  library(ggthemes) 
  library(ggrepel)
  library(ggforce)
  library(ggpmisc)
  library(ggprism)
  library(patchwork)
  library(extrafont)
  library(ggrepel)
  library(pheatmap)
  library(viridis)
  library(patchwork)
  library(fgsea)
  library(caret)
  library(msigdbr)
  library(matrixStats)
  library(NMF)
  library(rsample)
  library(vip)
  library(ROCR)
  library(glmnet)
  library(mlbench)
  library(tximeta)
  library(SummarizedExperiment)
  library(rjson)
  library(HDF5Array)
  library(DESeq2)
  # library(edgeR)
  library(ggpmisc)
  library(sva)
})
suppressMessages(loadfonts())
```
## setting ggplot theme
```{r setting ggplot theme}

base_size = 10

theme_set(theme_foundation(base_size = base_size,
                           base_family = 'Helvetica') + 
            theme(plot.title = element_blank(),
                panel.background = element_rect(colour = NA),
                plot.background = element_rect(colour = NA),
                panel.border = element_rect(colour = NA), 
                axis.line = element_line(), 
                axis.line.x = NULL, 
                axis.line.y = NULL, 
                axis.text = element_text(size = rel(0.95)), 
                axis.text.x = element_text(margin = margin(t = 0.8 * base_size/4)), 
                axis.text.x.top = element_text(margin = margin(b = 0.8 * base_size/4), vjust = 0), 
                axis.text.y = element_text(margin = margin(r = 0.5 * base_size/4), hjust = 1), 
                axis.text.y.right = element_text(margin = margin(l = 0.5 * base_size/4), hjust = 0), 
                axis.ticks = element_line(), 
                axis.ticks.length = unit(base_size/2.5, "pt"), axis.ticks.length.x = NULL, 
                axis.ticks.length.x.top = NULL, axis.ticks.length.x.bottom = NULL, 
                axis.ticks.length.y = NULL, axis.ticks.length.y.left = NULL, 
                axis.ticks.length.y.right = NULL,
                strip.text = element_text(size = 6),
                strip.background = element_blank(),
                legend.key.size= unit(0.08, "in"),
                legend.spacing = unit(0, "in"),
                legend.key = element_rect(colour = NA),
                legend.title = element_text(face="italic"),
                legend.text = element_text(face = 'bold'),
                legend.justification = c("right", "top"),
                legend.box.just = "right",
                legend.margin = margin(6, 6, 6, 6),
                plot.margin=margin(0.04,
                                   0.04,
                                   0.04,
                                   0.04,
                                   unit = "in"),
                # legend.margin = margin(-0.04,
                #                    -0.02,
                #                    0.08,
                #                    0.02,
                #                    unit = "in"),
                legend.box.spacing = unit(-0.02, 'in'),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank()
          ))

theme_set_fun <- function(){
  theme_set(theme_foundation(base_size = 8,
                           base_family = 'Helvetica') + 
            theme(plot.title = element_blank(),
                panel.background = element_rect(colour = NA),
                plot.background = element_rect(colour = NA),
                panel.border = element_rect(colour = NA),
                axis.title = element_text(face = "bold",size = rel(1)),
                axis.title.y = element_text(angle=90,vjust =2),
                axis.title.x = element_text(vjust = -0.2),
                axis.text = element_text(size = rel(1)), 
                axis.line = element_blank(),
                strip.text = element_text(size = 6),
                strip.background = element_blank(),
                legend.key.size= unit(0.08, "in"),
                legend.spacing = unit(0, "in"),
                legend.key = element_rect(colour = NA),
                legend.title = element_text(face="italic"),
                legend.text = element_text(face = 'bold'),
                plot.margin=margin(0.04,
                                   0.04,
                                   0.04,
                                   0.04,
                                   unit = "in"),
                legend.margin = margin(-0.04,
                                   -0.02,
                                   -0.08,
                                   0.02,
                                   unit = "in"),
                legend.box.spacing = unit(-0.02, 'in'),
                panel.grid.major = element_blank(),
                panel.grid.minor = element_blank()))
}

```
## figure dirs, and other globally relevant constants
```{r system I/O constants}
figure.dir <-  here('figures')
data.dir <-  here('data')
output.data.dir <- here('output.data')
reference.dir <- here('reference')

chr.order <- c('chr1', 'chr2', 'chr3', 
               'chr4', 'chr5', 'chr6', 
               'chr7', 'chr8', 'chr9',
               'chr10', 'chr11', 'chr12', 
               'chr13', 'chr14', 'chr15', 
               'chr16', 'chr17',
               'chr18', 'chr19', 'chr20', 
               'chr21', 'chr22', 'chrX', 'chrY')
```
## saving panels
```{r saving panel}
panel.figure.width = 8.7
panel.figure.height = 22.5

panel.save <- function(figure, name, width = 'single', height = 'single'){
  
  if (width == 'single'){
    panel.width = panel.figure.width
  } else if (width == 'medium'){
    panel.width = panel.figure.width*1.5
  } else{
    panel.width = panel.figure.width*2
  }
  
  if (height == 'single'){
    panel.height = panel.figure.height/3
  } else if (height == 'double'){
    panel.height = panel.figure.height/2
  } else{
    panel.height = panel.figure.height
  }
  
  ggsave(paste0(figure.dir, 
                'fig.',
                figure,
                '/', 
                name, 
                '.',
                panel.width, 
                'x', 
                panel.height, 
                '.pdf'),
       width = panel.width, 
       height = panel.height, 
       units = 'cm')
}

save_pheatmap_pdf <- function(x, filename, width=5, height=3) {
   stopifnot(!missing(x))
   stopifnot(!missing(filename))
   pdf(filename, width=width, height=height)
   grid::grid.newpage()
   grid::grid.draw(x$gtable)
   dev.off()
}
```
## gene sets for filtering, identification
```{r gene set I/O}
## sex linked genes to help clustering
gencode.sex.linked.genes <- 
  read_tsv(file.path(output.data.dir, 'reference.data/sex.linked.gene.list.tsv'),
           skip = 1,
           col_names = c('chr','ensg','gene'))

gen.34.lnc.names <- 
  read_table(file.path(output.data.dir, 'reference.data/gen.34.lncRNA.gene.names.txt'),
             col_names = 'gene')

gen.34.tx.2.gene <- 
  read_csv(file.path(output.data.dir, 'reference.data/gen.34.tx.to.gene.csv'),
             col_names = c('tx', 'gene')) %>% 
  separate(tx, sep = '\\|', into = c('enst','ensg',NA,NA,'tx','gene','len','biotype'))

gen.34.tx.2.gene %>% 
  select(-enst) %>% 
  #mutate(biotype = ifelse(biotype %in% c('Mt_rRNA', 'lncRNA', 'protein_coding'),
  #                        biotype, 'other')) %>% 
  group_by(biotype) %>% 
  tally() -> gen.34.tx.2.gene.biotype.count

gen.24.names <- 
  read_csv(file.path(output.data.dir, 'reference.data/gen.24.names.csv'),
           col_names = c('enst','ensg','tx','gene','len','biotype')) %>% 
  select(ensg, gene) %>% 
  distinct()
  
te.clades <- 
  read_csv(file.path(output.data.dir, 'reference.data','te.clades.csv'))

te.families.clades <- 
  read_csv(file.path(output.data.dir, 'reference.data','te.family.clade.csv'),
           col_names = c('gene', 'family', 'clade')) %>% 
  distinct()

gen.34.lnc.te.overlap <- 
  read_tsv(file.path(output.data.dir, 'reference.data','gen.34.lncRNA.exon.intersect.ucsc.gb.rep.txt'),
           col_names = c('chr','te.start','te.end','te','te.ol','strand',
                         'chr.2', 'exon.start', 'exon.end', 'enst', 'null', 'strand.2', 'exon.ol')) %>% 
  select(-null, -chr.2, -strand.2) %>% 
  merge(gen.34.tx.2.gene, by = 'enst') %>% 
  merge(te.families.clades, by.x = 'te', by.y = 'gene')

yang.ts.score.gtex <- 
  read_tsv(file.path(output.data.dir, 'reference.data', 'yang_et_al.ts_score_gencode.v19.tsv'),
           col_names = c('ensg','gene','tissue','tpm','group','ts_score'),
           skip = 1) %>% 
  filter(ts_score > 3) %>% 
  group_by(gene) %>% 
  filter(n() == 1) %>% 
  separate(tissue, sep = ' - ', into = c('tissue', NA))

gencode.35.tx.to.gene <- 
  read_csv(file.path(reference.dir, 'gencode.v35.tx.to.gene.csv'), 
                                  col_names = c('tx', 'gene')) %>% 
  separate(tx, sep = '\\|', into = c('enst','ensg',NA,NA,'tx','gene','len','biotype'))

gencode.35.tx.to.gene %>% 
  select(-enst) %>% 
  group_by(biotype) %>% 
  tally() -> gen.35.biotype.count

gencode.35.lnc.gene.names <- 
  read_csv(file.path(reference.dir, 'gencode.v35.lncRNA.gene.names.txt'), 
           col_names = 'gene')

ucsc.rmsk.tx.to.gene <-  
  read_csv(file.path(reference.dir, 'gencode.v35.ucsc.rmsk.tx.to.gene.csv'), 
                                  col_names = c('enst', 'gene')) %>%
  filter(grepl('^hg38', enst))

ucsc.rmsk.tx.to.gene %>% 
  mutate(tx = enst) %>% 
  separate(tx, sep = '_', into = c(NA, NA, 'gene', 'position', NA, NA, 'strand', NA)) %>% 
  mutate(position = sub('range=', '', position),
         strand = sub('strand=', '', strand)) %>% 
  merge(te.families.clades, by = 'gene') %>% 
  mutate(tmp.gene = gene,
         ensg = NA) %>% 
  select(enst, ensg, 'gene' = tmp.gene, 'len' = clade, 'biotype' = family) -> ucsc.rmsk.tx.to.gene.fmt


tot.gencode.35.tx.to.gene <- 
  read_csv(file.path(reference.dir, 'gencode.v35.tx.to.gene.csv'), 
                                  col_names = c('tx', 'gene')) %>% 
  mutate(enst = tx) %>% 
  separate(tx, sep = '\\|', into = c(NA,'ensg',NA,NA,'tx','gene','len','biotype'))

total.tx.to.gene <- 
  rbind(tot.gencode.35.tx.to.gene %>% select(enst, ensg, gene, len, biotype),
        ucsc.rmsk.tx.to.gene.fmt %>% select(enst = enst, ensg, gene, len, biotype))

total.tx.to.gene %>% 
  select(-enst) %>% 
  group_by(biotype) %>% 
  tally() -> total.tx.to.gene.count

g12c.marker.genes <- read_csv(file.path(output.data.dir, 'g12c.bulk.marker.genes.csv')) %>% 
  select(-X1)

g12c.sc.marker.genes <- read_csv(file.path(output.data.dir, 'g12c.scRNA.marker.genes.csv')) %>% 
  select(-X1)
```

# feature engineering:
## FUN: construct.quant.object
```{r}
load.tximeta.object <- function(reference){
  
  print(file.path(output.data.dir, paste0(reference, '_h5_se')))
  
  txi <- loadHDF5SummarizedExperiment(dir=file.path(output.data.dir, paste0(reference, '_h5_se')))
  gxi <- loadHDF5SummarizedExperiment(dir=file.path(output.data.dir, paste0(reference, '_gene_h5_se')))
  
  if (reference == 'process.aware.salmon'){
    
    gxi.split <- loadHDF5SummarizedExperiment(dir=file.path(output.data.dir, paste0(reference, '_gene_split_h5_se')))
    
    return(list('txi' = txi, 'gxi' = gxi, 'gxi.split' = gxi.split))
    
  } else {
    
    return(list('txi' = txi, 'gxi' = gxi))
  }
  
}

subset.tximeta.object <- function(tximeta.list, grep.string){
  
  if (grepl('\\|', grep.string)){
      
    subset.txi <- SummarizedExperiment::subset(tximeta.list$txi, 
                                         select = !grepl(grep.string, colData(tximeta.list$txi)$names))
  
    subset.gxi <- SummarizedExperiment::subset(tximeta.list$gxi, 
                                         select = !grepl(grep.string, colData(tximeta.list$gxi)$names))
  } else{
        
    subset.txi <- SummarizedExperiment::subset(tximeta.list$txi, 
                                         select = grepl(grep.string, colData(tximeta.list$txi)$names))
  
    subset.gxi <- SummarizedExperiment::subset(tximeta.list$gxi, 
                                         select = grepl(grep.string, colData(tximeta.list$gxi)$names))
  }
  
  return(list('txi' = subset.txi,
              'gxi' = subset.gxi))
}
```
### CALL: construct.quant.object
```{r}
salmon.quant <- load.tximeta.object('salmon')
ucsc.rmsk.salmon.quant <- load.tximeta.object('ucsc.rmsk.salmon')
process.aware.salmon.quant <- load.tximeta.object('process.aware.salmon')

aale.salmon.quant <- load.tximeta.object('aale_salmon')

salmon.quant$exoTIC_2D <- subset.tximeta.object(salmon.quant, 'exoTIC.2D')
salmon.quant$rnaEasy_2D <- subset.tximeta.object(salmon.quant, 'rnaEASY.2D')
salmon.quant$exoTIC_3D <- subset.tximeta.object(salmon.quant, 'exoTIC.3D')
salmon.quant$rnaEasy_3D <- subset.tximeta.object(salmon.quant, 'rnaEASY.3D')

salmon.quant$rnaEasy_dmso <- subset.tximeta.object(salmon.quant, 'AMG|exoTIC')

salmon.quant$exoTIC_all <- subset.tximeta.object(salmon.quant, 'exoTIC')
salmon.quant$rnaEasy_all <- subset.tximeta.object(salmon.quant, 'rnaEASY')

salmon.quant$two_D <- subset.tximeta.object(salmon.quant, '2D')
salmon.quant$three_D <- subset.tximeta.object(salmon.quant, '3D')

ucsc.rmsk.salmon.quant$exoTIC_2D <- subset.tximeta.object(ucsc.rmsk.salmon.quant, 'exoTIC.2D')
ucsc.rmsk.salmon.quant$rnaEasy_2D <- subset.tximeta.object(ucsc.rmsk.salmon.quant, 'rnaEASY.2D')
ucsc.rmsk.salmon.quant$exoTIC_3D <- subset.tximeta.object(ucsc.rmsk.salmon.quant, 'exoTIC.3D')
ucsc.rmsk.salmon.quant$rnaEasy_3D <- subset.tximeta.object(ucsc.rmsk.salmon.quant, 'rnaEASY.3D')

ucsc.rmsk.salmon.quant$exoTIC_all <- subset.tximeta.object(ucsc.rmsk.salmon.quant, 'exoTIC')
ucsc.rmsk.salmon.quant$rnaEasy_all <- subset.tximeta.object(ucsc.rmsk.salmon.quant, 'rnaEASY')

process.aware.salmon.quant$exoTIC_2D <- subset.tximeta.object(process.aware.salmon.quant, 'exoTIC.2D')
process.aware.salmon.quant$rnaEasy_2D <- subset.tximeta.object(process.aware.salmon.quant, 'rnaEASY.2D')
process.aware.salmon.quant$exoTIC_3D <- subset.tximeta.object(process.aware.salmon.quant, 'exoTIC.3D')
process.aware.salmon.quant$rnaEasy_3D <- subset.tximeta.object(process.aware.salmon.quant, 'rnaEASY.3D')

```
### CALL: build nanosight
```{r}

read_csv(file.path(output.data.dir, 'exoTIC_3D_nano.csv')) %>%  
    gather(sample, value, -size) %>% 
    mutate(condition = sample) %>% 
    separate(condition, sep = '_', into = c('context', 'platform', 'condition')) %>% 
    group_by(size, condition) %>%
    mutate(mean = (mean(value))/1e6,
         sd = (sd(value))/1e6) -> exoTIC.3D.nanosight.df

head(read_csv(file.path(output.data.dir, 'exoTIC_2D_nano.csv')), -1) %>% 
    gather(sample, value, -size) %>% 
    mutate(condition = sample) %>% 
    separate(condition, sep = '_', into = c('context', 'platform', 'condition')) %>% 
    group_by(size, condition) %>%
    mutate(mean = (mean(value))/1e6,
         sd = (sd(value))/1e6) %>% 
  drop_na() -> exoTIC.2D.nanosight.df

read_csv(file.path(output.data.dir, 'rnaEasy_AMG_nano.csv')) %>% 
  gather(sample, value, -size) %>% 
  mutate(condition = sample) %>% 
  separate(condition, sep = '_', into = c('context', 'platform', 'dose', 'condition')) %>% 
  mutate(value = value/1e6) %>% 
  drop_na() -> rnaEasy.amg.nanosight.df

read_csv(file.path(output.data.dir, 'rnaEasy_DMSO_nano.csv')) %>% 
  select(-X2) %>% 
  gather(sample, value, -size) %>%
  mutate(condition = sample) %>% 
  separate(condition, sep = '_', into = c('context', 'platform', 'condition', NA)) %>% 
  group_by(size, context) %>%
  mutate(mean = (mean(value))/1e6,
         sd = (sd(value))/1e6) -> rnaEasy.dmso.nanosight.df

# nanosight.df %>% 
#   filter(sample == '2d_rnaEasy_NA_DMSO') %>% 
#   group_by(sample) %>% 
#   mutate(weight = sum(size*value)/sum(value)) %>% 
#   top_n(1, weight) #%>% 
#   select(-size, -value) #%>% filter(sample == '2d_rnaEasy_NA_DMSO')
# distinct() %>% 
#   filter(!condition %in% c('0.05', '0.01')) %>% 
#   ungroup() %>% 
#   mutate(condition = ifelse(condition == '0.1', 'amg', condition),
#          context = toupper(context), toupper(condition), 
#          platform = ifelse(platform == 'exotic', 'exoTIC', 'rnaEASY'))



```
## metadata
```{r}
colData(salmon.quant$gxi) %>% 
  as.data.frame() %>% 
  remove_rownames() %>%
  mutate(condition = names) %>% 
  separate(condition, sep = '[.]', into = c('platform', 'context', NA, 'condition')) %>% 
  mutate(batch = ifelse(grepl('exoTIC.3D.*.DMSO', names), 1, 0)) -> metadata.df
```
## FUN: extract.process.aware
```{r}
txome.json <- file.path(reference.dir, 'gencode.v35.annotation.expanded.json')
txome.tsv <- file.path(reference.dir, 'gencode.v35.annotation.expanded.tsv')


txi.iranges <- rowRanges(process.aware.salmon.quant$gxi.split)

split.gxi <- process.aware.salmon.quant$gxi.split

assay(split.gxi, 'unspliced') %>% 
  as.data.frame() %>% 
  rownames_to_column('gene') %>% 
  gather(sample, count, -gene) %>% 
  filter(!grepl('SRR|luad', sample)) %>% 
  group_by(sample) %>% 
  summarize(unsplice_count = sum(count)) -> unsplice.df

assay(split.gxi, 'spliced') %>% 
  as.data.frame() %>% 
  rownames_to_column('gene') %>% 
  gather(sample, count, -gene) %>% 
  filter(!grepl('SRR|luad', sample)) %>% 
  group_by(sample) %>% 
  summarize(splice_count = sum(count)) -> splice.df

colnames(split.gxi) %>% 
  as.data.frame() %>% 
  cbind(as.data.frame(metadata(split.gxi)$quantInfo$percent_mapped)) %>% 
  select('sample' = '.', 'mapping_rate' = 'metadata(split.gxi)$quantInfo$percent_mapped') -> mapping_rate_df

merge(unsplice.df, splice.df, by = 'sample') %>% 
  mutate(unsplice_rate = unsplice_count / (unsplice_count + splice_count)) %>% 
  merge(mapping_rate_df, by = 'sample') -> unsplice.df 

unsplice.df %>% 
  separate(sample, sep = '[.]', into = c('platform', 'context', 'rep', 'condition')) %>%
  ggplot(aes(reorder(platform, -unsplice_rate), unsplice_rate, color = context)) +
  geom_boxplot(fill = NA) +
  geom_sina()

```
### CALL: combat.batch effect
```{r}
metadata.df %>% mutate_if(is.character, as.factor)-> pheno

model.matrix(~as.factor(batch), data = pheno) -> mod
model.matrix(~1, data = pheno) -> mod_0

edata <- data.matrix(assay(salmon.quant$gxi, 'counts') %>%
                      as.data.frame() %>%
                      rownames_to_column('gene') %>%
                      filter_at(vars(contains('.')), all_vars(. >= 5)) %>%
                      select(-gene)) %>% as.matrix()

n.sv <- num.sv(edata, mod, method = 'leek')
n.sv

svobj <- sva(edata, mod, mod_0, n.sv=6)
svobj$sv

pValues = f.pvalue(edata,mod,mod_0)
qValues = p.adjust(pValues,method="BH")

modSv = cbind(mod,svobj$sv)
mod0Sv = cbind(mod_0,svobj$sv)
pValuesSv = f.pvalue(edata,modSv,mod0Sv)
qValuesSv = p.adjust(pValuesSv,method="BH")
qValuesSv
pValuesSv
modSv

cbind(metadata.df, modSv) %>% rename('names' = 'sample') -> batch.annotation
```
## FUN: plot tx count
```{r}
tx.ome.cov.plt <- function(tx.counts){
  tx.counts %>% 
    rownames_to_column('enst') %>% 
    merge(gencode.35.tx.to.gene, by = 'enst') %>% 
    gather(names, count, -enst, -ensg:-biotype) %>% 
    merge(metadata.df, by = 'names') %>% 
    filter(biotype != 'Mt_rRNA') %>% 
    filter(count >= 5) %>% 
    group_by(biotype, names, condition) %>% 
    mutate(biotype.count = n()) %>% 
    select(biotype, names, condition, biotype.count) %>% 
    distinct() %>% 
    merge(gen.35.biotype.count, by = 'biotype') %>% 
    mutate(coverage = biotype.count/n) 
}
```

## FUN: run.de.seq
```{r}
run.de.seq <- function(tximeta.object){
  
  tximeta.name <- deparse(substitute(tximeta.object))
  
  print(tximeta.name)
  
  colData(tximeta.object$gxi) %>% 
    as.data.frame() %>% 
    mutate(sample = names) %>% 
    separate(names, sep = '[.]', into = c(NA, 'context', NA, NA)) -> colData.gxi
  
  print(colData.gxi)

  count.matrix.gxi <- assay(tximeta.object$gxi, 'counts') %>% 
    as.data.frame() %>% 
    rownames_to_column('gene') %>% 
    mutate_if(is.numeric, round) %>% 
    column_to_rownames('gene')
    
  colData(tximeta.object$txi) %>% 
    as.data.frame() %>% 
    mutate(sample = names) %>% 
    separate(names, sep = '[.]', into = c(NA, 'context', NA, NA)) -> colData.txi

  count.matrix.txi <- assay(tximeta.object$txi, 'counts') %>% 
    as.data.frame() %>% 
    rownames_to_column('tmp') %>% 
    mutate_if(is.numeric, round) %>% 
    column_to_rownames('tmp')
  
  
  if (grepl('all', tximeta.name)){
    print('2D vs 3D')
    dds.gxi <- DESeqDataSetFromMatrix(countData = count.matrix.gxi, 
                                      colData = colData.gxi, 
                                      design = as.formula(~context + condition:context))
    
  } else if (grepl('_D', tximeta.name)){
    dds.gxi <- DESeqDataSetFromMatrix(countData = count.matrix.gxi, 
                                      colData = colData.gxi, 
                                      design = as.formula(~condition + condition:platform))
  } else if (grepl('_dmso', tximeta.name)){
    dds.gxi <- DESeqDataSetFromMatrix(countData = count.matrix.gxi, 
                                      colData = colData.gxi, 
                                      design = as.formula(~context))
    dds.txi <- DESeqDataSetFromMatrix(countData = count.matrix.txi, 
                                    colData = colData.txi, 
                                    design = as.formula(~context))
    dds.txi <- estimateSizeFactors(dds.txi)
  } else{
    dds.gxi <- DESeqDataSetFromMatrix(countData = count.matrix.gxi, 
                                      colData = colData.gxi, 
                                      design = as.formula(~condition))
  }
  
  if (!exists('dds.txi')){
  dds.txi <- DESeqDataSetFromMatrix(countData = count.matrix.txi, 
                                    colData = colData.txi, 
                                    design = as.formula(~condition))
  }
  
  dds.txi <- estimateSizeFactors(dds.txi)
  
  dds.txi.norm.counts <- as.data.frame(counts(dds.txi, normalized=T))
  
  dds.txi.vst.counts <- as.data.frame(assay(vst(dds.txi, blind = F)))
  
  dds.gxi <- estimateSizeFactors(dds.gxi)
  
  dds.gxi.norm.counts <- as.data.frame(counts(dds.gxi, normalized=T))
  
  dds.gxi.vst.counts <-  as.data.frame(assay(vst(dds.gxi, blind = F)))
  
  if (!grepl('exoTIC|rnaEasy|_D|aale', tximeta.name)){
    print('alt approach')
      if (grepl('ucsc.rmsk', tximeta.name)){  
    
        print('rmsk')
    
        dds.gxi.norm.counts <- dds.gxi.norm.counts %>% 
          rownames_to_column('gene') 
        
        return(list('counts' = dds.gxi.norm.counts ,
                    'txi.counts' = dds.txi.norm.counts))
        
        quit()
      }
    
      return(list('counts' = dds.gxi.norm.counts %>% 
                    rownames_to_column('ensg') %>% 
                    merge(gencode.35.tx.to.gene %>% 
                            select(gene, ensg), by = 'ensg') %>% 
                    distinct(), 
                  'txi.counts' = dds.txi.norm.counts))
      quit()

  }
  
  filter <- rowSums(counts(dds.gxi, normalized=T) >= 10) >= 2
  
  dds.gxi <- DESeq(dds.gxi[filter,])
  
  print(resultsNames(dds.gxi))
  
  print(tail(resultsNames(dds.gxi), n=1))
  
  if (grepl('ucsc.rmsk', tximeta.name)){  
    
    print('rmsk')
    
    dds.gxi.lfc <- lfcShrink(dds.gxi, coef = tail(resultsNames(dds.gxi), n=1)) %>% 
      as.data.frame() %>% 
      filter(padj <= 0.05) %>% 
      rownames_to_column('gene') %>% 
      select(gene, baseMean, log2FoldChange, padj)
    
    dds.gxi.norm.counts <- dds.gxi.norm.counts %>% 
      rownames_to_column('gene') 
    
  } else{
    
    dds.gxi.lfc <- lfcShrink(dds.gxi, coef = tail(resultsNames(dds.gxi), n=1)) %>% 
      as.data.frame() %>% 
      filter(padj <= 0.05) %>% 
      rownames_to_column('ensg') %>% 
      merge(gencode.35.tx.to.gene %>% select(gene, ensg) %>% distinct(), by = 'ensg') %>%
      select(gene, baseMean, log2FoldChange, padj)
    
    plotMA(lfcShrink(dds.gxi, coef = tail(resultsNames(dds.gxi), n=1)))
    
    dds.gxi.norm.counts <- dds.gxi.norm.counts %>% 
      rownames_to_column('ensg') %>% 
      merge(gencode.35.tx.to.gene %>% select(ensg, gene) %>% distinct(), by = 'ensg') %>% 
      distinct() %>% 
      select(ensg, gene, everything())
  }

return(list('de' = dds.gxi.lfc, 
            'counts' = dds.gxi.norm.counts, 
            'txi.counts' = dds.txi.norm.counts, 
            'txi.vst.counts' = dds.txi.vst.counts,
            'gxi.vst.counts' = dds.gxi.vst.counts))
  
}

```
## CALL: run.de.seq
```{r}
salmon.quant$exoTIC_2D$deseq <- run.de.seq(salmon.quant$exoTIC_2D)
salmon.quant$rnaEasy_2D$deseq <- run.de.seq(salmon.quant$rnaEasy_2D)
salmon.quant$exoTIC_3D$deseq <- run.de.seq(salmon.quant$exoTIC_3D)
salmon.quant$rnaEasy_3D$deseq <- run.de.seq(salmon.quant$rnaEasy_3D)
salmon.quant$exoTIC_all$deseq <- run.de.seq(salmon.quant$exoTIC_all)
salmon.quant$rnaEasy_all$deseq <- run.de.seq(salmon.quant$rnaEasy_all)
salmon.quant$two_D$deseq <- run.de.seq(salmon.quant$two_D)
salmon.quant$three_D$deseq <- run.de.seq(salmon.quant$three_D)
salmon.quant$deseq <- run.de.seq(salmon.quant)

salmon.quant$rnaEasy_dmso$deseq <- run.de.seq(salmon.quant$rnaEasy_dmso)

aale.salmon.quant$deseq <- run.de.seq(aale.salmon.quant)

# salmon.quant$rnaEasy_3D$deseq$de %>% 
#   mutate(log2FoldChange = -log2FoldChange) ->salmon.quant$rnaEasy_3D$deseq$de
# salmon.quant$rnaEasy_2D$deseq$de %>% 
#   mutate(log2FoldChange = -log2FoldChange) ->salmon.quant$rnaEasy_2D$deseq$de
# salmon.quant$exoTIC_2D$deseq$de %>% 
#   mutate(log2FoldChange = -log2FoldChange) ->salmon.quant$exoTIC_2D$deseq$de
# salmon.quant$exoTIC_3D$deseq$de %>% 
#   mutate(log2FoldChange = -log2FoldChange) ->salmon.quant$exoTIC_3D$deseq$de

ucsc.rmsk.salmon.quant$exoTIC_2D$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$exoTIC_2D)
ucsc.rmsk.salmon.quant$rnaEasy_2D$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$rnaEasy_2D)
ucsc.rmsk.salmon.quant$exoTIC_3D$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$exoTIC_3D)
ucsc.rmsk.salmon.quant$rnaEasy_3D$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$rnaEasy_3D)
ucsc.rmsk.salmon.quant$exoTIC_all$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$exoTIC_all)
ucsc.rmsk.salmon.quant$rnaEasy_all$deseq <- run.de.seq(ucsc.rmsk.salmon.quant$rnaEasy_all)
ucsc.rmsk.salmon.quant$deseq <- run.de.seq(ucsc.rmsk.salmon.quant)
# 
# ucsc.rmsk.salmon.quant$rnaEasy_3D$deseq$de %>% 
#   mutate(log2FoldChange = -log2FoldChange) ->ucsc.rmsk.salmon.quant$rnaEasy_3D$deseq$de
# ucsc.rmsk.salmon.quant$rnaEasy_2D$deseq$de %>% 
#   mutate(log2FoldChange = -log2FoldChange) ->ucsc.rmsk.salmon.quant$rnaEasy_2D$deseq$de
# ucsc.rmsk.salmon.quant$exoTIC_2D$deseq$de %>% 
#   mutate(log2FoldChange = -log2FoldChange) ->ucsc.rmsk.salmon.quant$exoTIC_2D$deseq$de
# ucsc.rmsk.salmon.quant$exoTIC_3D$deseq$de %>% 
#   mutate(log2FoldChange = -log2FoldChange) ->ucsc.rmsk.salmon.quant$exoTIC_3D$deseq$de


```
## CALL: tmm.norm
```{r}

salmon.quant$rnaEasy_all$tmm.counts <- 
  edgeR::calcNormFactors(assay(salmon.quant$rnaEasy_all$gxi, 'counts') %>% 
    as.data.frame() %>% 
    rownames_to_column('gene') %>% 
    mutate_if(is.numeric, round) %>% 
    column_to_rownames('gene'), method = 'TMM')

edgeR::cp
```

## FUN: elbow.plt
```{r}
elbow.plt <- function(df){
  k.max <- nrow(df) - 1
  k.values <- 1:k.max
  
  cluster.compute <- function(k){
    cluster.out <- kmeans(df, 
                          k, 
                          nstart = 25)
    return(cluster.out$tot.withinss)
  }
  
  elbow.df <- tibble()
  for(i in k.values){
    print(i)
    temp.df <- 
      tibble(k = i, 
             ss = cluster.compute(i))
    elbow.df <- rbind(temp.df, elbow.df)
  }
  
  ggplot(elbow.df,
         aes(k, ss)) + 
    geom_point() +
    geom_line() + 
    geom_vline(xintercept = k.values, 
               linetype = 'dashed', 
               alpha = 0.3)
}
```
## FUN: make.fgsea.ranks
```{r}
make.fgsea.ranks <- function(de.seq){
  
  # build a ranked gene list in the appropriate format (named list)
  # using shrunken log2 Fold Change values from DESeqII
  de.seq <- 
    de.seq %>%
    mutate(rank = as.double(log2FoldChange),
           Gene = as.character(gene)) %>% 
    select(Gene, rank) %>%
    mutate(rank = scale(rank)) 
  
  de.seq.rnk <- 
    de.seq$rank
  de.seq.rnk <- 
    setNames(de.seq$rank, de.seq$Gene)

  # run the GSEA implementation on the hallmark sets supplemented with custom sets
  print('fgsea')
  de.seq.fgsea.res <- 
    fgsea(pathways = msig.df, 
          stats = de.seq.rnk, eps = 0.0)
  
  print('padj filter and clean names')
  # convert to a tidy format, clean the names for display
  de.seq.fgsea.df <-
    as_tibble(de.seq.fgsea.res) %>%
    filter(padj <= 0.05) %>%
    mutate(pathway = gsub('_', ' ', pathway))

  
  return(de.seq.fgsea.df)
  # clean tmp objects
  rm(de.seq, de.seq.rnk, de.seq.fgsea.res, de.seq.fgsea.df)
  
}
```
### CALL: make.fgsea.ranks
```{r}
msig.df <- 
  msigdbr(species = 'Homo sapiens', category = 'H') %>% 
  select(gene_symbol, gs_name) %>% 
  split(x = .$gene_symbol, f = .$gs_name)

salmon.quant$exoTIC_2D$deseq$c6.fgsea <- make.fgsea.ranks(salmon.quant$exoTIC_2D$deseq$de)
salmon.quant$rnaEasy_2D$deseq$c6.fgsea <- make.fgsea.ranks(salmon.quant$rnaEasy_2D$deseq$de)
salmon.quant$rnaEasy_3D$deseq$c6.fgsea <- make.fgsea.ranks(salmon.quant$rnaEasy_3D$deseq$de)


salmon.quant$rnaEasy_dmso$deseq$h.fgsea <- make.fgsea.ranks(salmon.quant$rnaEasy_dmso$deseq$de)

salmon.quant$exoTIC_3D$deseq$fgsea <- make.fgsea.ranks(salmon.quant$exoTIC_3D$deseq$de)
salmon.quant$rnaEasy_all$deseq$fgsea <- make.fgsea.ranks(salmon.quant$rnaEasy_all$deseq$de)
salmon.quant$exoTIC_all$deseq$fgsea <- make.fgsea.ranks(salmon.quant$exoTIC_all$deseq$de)

salmon.quant$two_D$deseq$fgsea <- make.fgsea.ranks(salmon.quant$two_D$deseq$de)
salmon.quant$three_D$deseq$fgsea <- make.fgsea.ranks(salmon.quant$three_D$deseq$de)

salmon.quant$exoTIC_2D$deseq$fgsea %>%
  select(pathway, padj, NES) %>% 
  mutate(experiment = 'exoTIC_2D') %>% 
  rbind(salmon.quant$rnaEasy_2D$deseq$fgsea %>% 
  select(pathway, padj, NES) %>% 
  mutate(experiment = 'rnaEasy_2D')) %>% 
  rbind(salmon.quant$rnaEasy_3D$deseq$fgsea %>% 
  select(pathway, padj, NES) %>% 
  mutate(experiment = 'rnaEasy_3D')) %>% 
  # select(-NES) %>% 
  mutate(padj = -log10(padj)) %>% 
  separate(pathway, sep = 'HALLMARK ', into = c(NA, 'pathway')) %>% 
  select(-padj) %>% 
  spread(experiment, NES) %>%
  replace(is.na(.), 0) %>%
  column_to_rownames('pathway') %>%
  pheatmap(color = viridis(32, option = 'magma'))
  # ggplot(aes(pathway,NES, color = padj)) +
  # geom_boxplot(fill = NA) +
  # geom_sina() + theme(axis.text.x = element_text(angle = 40, hjust = 1))

salmon.quant$exoTIC_2D$deseq$c6.fgsea %>%
  select(pathway, padj, NES) %>% 
  rename(NES = 'exoTIC_2D') %>% 
  merge(salmon.quant$rnaEasy_2D$deseq$c6.fgsea %>% 
  select(pathway, padj, NES) %>% 
  rename(NES = 'rnaEasy_2D'), by = 'pathway') %>% 
  merge(salmon.quant$rnaEasy_3D$deseq$c6.fgsea %>%
  select(pathway, padj, NES) %>%
  rename(NES = 'rnaEasy_3D'), by = 'pathway') %>%
  column_to_rownames('pathway') %>%
  select(-contains('padj')) %>% 
  pheatmap()#color = viridis(32, option = 'magma'))

salmon.quant$exoTIC_2D$deseq$fgsea %>% 
  filter(pathway == 'HALLMARK MYC TARGETS V1') %>% 
  select(leadingEdge) %>% 
  unlist() %>% 
  unname() -> myc.leading.edge

salmon.quant$rnaEasy_3D$deseq$fgsea %>% 
  filter(pathway == 'ZWANG TRANSIENTLY UP BY 2ND EGF PULSE ONLY') %>% 
  select(leadingEdge) %>% 
  unlist() %>% 
  unname() -> egf.leading.edge

salmon.quant$rnaEasy_2D$deseq$fgsea %>% 
  filter(pathway == 'RODRIGUES THYROID CARCINOMA POORLY DIFFERENTIATED DN') %>% 
  select(leadingEdge) %>% 
  unlist() %>% 
  unname() -> thy.carcinoma.dn.leading.edge

salmon.quant$exoTIC_2D$deseq$fgsea %>% 
  filter(pathway == 'MANNO MIDBRAIN NEUROTYPES HGABA') %>% 
  select(leadingEdge) %>% 
  unlist() %>% 
  unname() -> midbrain.neuro.leading.edge

hallmark.kras.up <- msig.df$HALLMARK_KRAS_SIGNALING_UP

hallmark.emt <- msig.df$HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION

salmon.quant$rnaEasy_3D$deseq$fgsea %>%
  merge(salmon.quant$rnaEasy_2D$deseq$fgsea, by = 'pathway') %>%
  select(pathway, NES.x, NES.y) %>%
  ggplot(aes(NES.x, NES.y)) +
  geom_point() +
  geom_text_repel(aes(label = pathway), size = rel(2)) + 
  geom_hline(yintercept = 0, linetype = 'dashed', alpha=0.2) + 
  geom_vline(xintercept = 0, linetype = 'dashed', alpha=0.2)

```
## FUN: rerwip create.tcga.obj
```{r}

if (! file.exists(file.path(output.data.dir, 'tcga.data', 'luad.data.frame.tsv'))){
    
  toil.survival <- read_tsv(file.path(output.data.dir, 'tcga.data/TCGA-LUAD.survival.tsv.gz')) 
  
  luad.phenotypes <- read_tsv(file.path(output.data.dir, 'tcga.data/TCGA-LUAD.GDC_phenotype.tsv.gz')) %>% 
    select(sample = submitter_id.samples, 
           sample_type = sample_type.samples,
           age = age_at_index.demographic,
           sex = gender.demographic,
           race = race.demographic)
  
  luad.mutect <- read_tsv(file.path(output.data.dir, 'tcga.data/TCGA-LUAD.mutect2_snv.tsv.gz')) %>% 
    filter(gene %in% c('KRAS', 'TP53')) %>% 
    select(sample = Sample_ID, gene, aa_change = Amino_Acid_Change) %>% 
    pivot_wider(names_from = gene, values_from = aa_change)
  
  luad.phenotypes.merge <- luad.phenotypes %>% 
    merge(luad.mutect, by = 'sample', all.x = T) %>% 
    filter(sample_type %in% c('Primary Tumor', 'Solid Tissue Normal')) %>% 
    merge(toil.survival, by = 'sample') %>% 
    mutate(sample = str_sub(sample, end = str_length(sample) - 1))
  
  toil.norm.counts <- read_csv(file.path(output.data.dir, 'tcga.data/toil.counts.csv'))
  
  norm.counts.tidy <- 
    toil.norm.counts %>% 
    select('ensg' = gene, contains('TCGA')) %>% 
    discard(~sum(.x == 0)/length(.x) >= 0.8) %>% 
    gather(sample, count, -ensg) 
  
  rm(toil.norm.counts)
  
  norm.counts.tidy.filt <- 
    norm.counts.tidy %>% 
    filter(sample %in% luad.phenotypes.merge$sample)
  
  norm.counts.tidy.filt.merge <- 
    norm.counts.tidy.filt %>% 
    merge(gen.24.names, by = 'ensg') %>%
    select(-ensg)
  
  norm.counts.final <- 
    norm.counts.tidy.filt.merge %>% 
    merge(luad.phenotypes.merge, by = 'sample') %>% 
    distinct()
  
  write_tsv(norm.counts.final %>% 
    mutate(KRAS = paste(unlist(KRAS[1]), sep = ',', collapse = ', '),
           TP53 = paste(unlist(TP53[1]), sep = ',', collapse = ', ')), 
    file = file.path(output.data.dir, 'tcga.data', 'luad.data.frame.tsv'))
  
}


norm.counts.final %>% 
  filter(gene %in% g12c.sc.marker.genes$`G12C induced`) %>% 
  select(sample, count, gene) %>% 
  pivot_wider(names_from = gene, values_from = count, values_fn = mean) %>% 
  column_to_rownames('sample') %>% 
  pheatmap(show_rownames = F, show_colnames = F, 
           annotation_row = luad.phenotypes.merge %>% 
             select(sample, sample_type, KRAS) %>% 
             mutate(KRAS = paste(KRAS)) %>% 
             filter(! sample %in% c('TCGA-44-2656-01', 'TCGA-44-2668-01', 
                                    'TCGA-44-3917-01', 'TCGA-44-3918-01', 'TCGA-44-5645-01', 'TCGA-44-6775-01')) %>% 
             distinct() %>% 
             column_to_rownames('sample'))

luad.phenotypes.merge %>% 
   select(sample, sample_type, KRAS) %>% 
   mutate(KRAS = paste(KRAS)) -> tmp.df




# toil.quant$myc.avg.counts.tidy.stratified <-
#   toil.quant$norm.counts.tidy %>% 
#   filter(type == 'Primary Tumor') %>% 
#   filter(gene %in% myc.list) %>% 
#   group_by(sample, category, disease, tissue, type) %>% 
#   summarize(avg.count = mean(count)) %>% 
#   mutate(gene.set = 'myc') %>% 
#   group_by(gene.set) %>% 
#   mutate(stratum = ifelse(avg.count >= quantile(avg.count, 0.66),
#                           'top-third', ifelse(avg.count <= quantile(avg.count, 0.33),
#                                               'bottom-third', 'middle'))) %>% 
#   filter(stratum != 'middle') %>% 
#   spread(gene.set, stratum)

# toil.quant$norm.counts.tidy.stratified <- 
#   toil.quant$norm.counts.tidy %>% 
#   filter(!grepl('GTEX', sample)) %>% 
#   group_by(gene) %>%
#   mutate(stratum = ifelse(count >= quantile(count, 0.66),
#                           'top-third', ifelse(count <= quantile(count, 0.33),
#                                               'bottom-third', 'middle')))


library(survival)
library(survminer)

toil.quant$multi.gene.de.set <-
  toil.quant$norm.counts.tidy.stratified %>%
  select(-ensg, -count, -category, -disease, -tissue,
         -study) %>% 
  #filter(gene %in% filter(panc1.intra.quant$de.seq$gencode.34.salmon.out, log2FoldChange < -1)$gene) %>%
  # filter(gene %in% c('KRAS', "GRAP2", "PDLIM1", "ACRBP", "PTPN18", "MAX")) %>%
  filter(gene %in% c('IFI27', 'GIMAP4', 'CARD16', 'CCL14', 'LY96')) %>%
  # filter(gene %in% c(patient.plasma.quant$rfe.condition.profile$optVariables, 'MIR4435-2HG')) %>%
  mutate(stratum = relevel(as.factor(stratum), ref = 'middle')) %>%
  distinct() %>% 
  pivot_wider(id_cols = sample, names_from = gene, values_from = stratum) %>% 
  mutate_if(is.list, as.factor)

surv.obj <- Surv(time = filter(toil.quant$toil.survival,
                                          sample %in% toil.quant$multi.gene.de.set$sample)$OS.time,
                            event = filter(toil.quant$toil.survival,
                                           sample %in% toil.quant$multi.gene.de.set$sample)$OS)

surv.fit <- survfit(as.formula(surv.obj ~ IFI27), data = toil.quant$multi.gene.de.set)
ggsurvplot(surv.fit, pval = T)
coxph.fit <- coxph(surv.obj ~ GIMAP4 + IFI27 + CARD16 + CCL14 + LY96,
                   data = toil.quant$multi.gene.de.set)
ggforest(coxph.fit) + theme_set_fun()

```
### classify tcga on plasma features
```{r}
set.seed(456)

opt.var.toil.count <- toil.quant$norm.counts.tidy.z.scores %>% 
  filter(gene %in% patient.plasma.quant$rfe.condition.profile$optVariables) %>% 
  # filter(gene %in% c('KRAS', "ETS1", "FOS", 
  #                    "JUNB", "NFKB1", "ETV4",
  #                    "LINC01670", "LINC00294","AC010531.1")) %>% 
  select(gene, sample, count, type) %>%
  mutate(count = log(count + 1)) %>% 
  spread(gene, count) %>% 
  #filter(type %in% c('Primary Tumor', 'Normal Tissue')) %>% 
  mutate(type = sub(' ', '_', type)) %>% 
  column_to_rownames('sample')

split_idx <- createDataPartition(opt.var.toil.count$type, p = 0.75, list = F)

split_train <- opt.var.toil.count[split_idx, ]
split_test <- opt.var.toil.count[-split_idx, ]

default_glm_mod = train(
  form = type ~ .,
  data = split_train,
  trControl = trainControl(method = "cv", number = 5,
                           savePredictions = T, classProbs = T),
  method = "glm",
  family = "binomial", 
  maxit = 150
)

calc_acc = function(actual, predicted) {
  print(paste0('Actual - Predicted (Accuracy) = ', mean(actual == predicted)))
}

calc_acc(actual = split_test$type,
         predicted = predict(default_glm_mod, newdata = split_test))

confusionMatrix(predict(default_glm_mod, newdata = split_test, type = 'prob'), )



plot.roc(default_glm_mod$pred$obs, default_glm_mod$pred$Primary_Tumor)

ggplot(default_glm_mod$pred, 
       aes(m = Primary_Tumor, d = factor(obs, levels = c("Normal_Tissue", "Primary_Tumor")))) + 
    geom_roc(hjust = -0.4, vjust = 1.5) + xlim(0,0.25)

summary(default_glm_mod)

predict(default_glm_mod, newdata = split_test, type = 'prob')

tot.var.count <- toil.quant$norm.counts.tidy.z.scores %>% 
  filter(gene %in% patient.plasma.quant$de.seq$gencode.34.salmon.out$gene) %>% 
  select(gene, sample, count, type) %>%
  mutate(count = log(count + 1)) %>% 
  spread(gene, count) %>% 
  filter(type %in% c('Primary Tumor', 'Normal Tissue')) %>% 
  column_to_rownames('sample')

tot.var.split <- createDataPartition(tot.var.count$type, p = 0.75, list = F)
tot.var_train <- opt.var.toil.count[tot.var.split, ]
tot.var_test <- opt.var.toil.count[-tot.var.split, ]

tot.var.x <- tot.var_train %>% select(-type) %>% as.matrix()
tot.var.y <- tot.var_train %>% select(type) %>% mutate(type = ifelse(grepl('Normal', type), 0, 1)) %>% as.matrix()

type <- 'auc'
cv.fit <- cv.glmnet(tot.var.x, tot.var.y, family="binomial", type.measure = type, keep = TRUE)
plot(cv.fit)

min.ind <- which(cv.fit$lambda == cv.fit$lambda.min)
pred <- expit(cv.fit$fit.preval[,min.ind])

fitted.roc <- roc(tot.var.y,pred)
plot(fitted.roc)

lm(tot.var.y~pred-1)

```
### boehmke classifier
```{r}
set.seed(123)

salmon.quant$panc_v_ctrl$rfe.condition.profile$optVariables %>% 
  as.tibble() %>% 
  select(ensg = value) %>% 
  merge(gencode.35.tx.to.gene %>% select(ensg, gene), by = 'ensg') %>% 
  distinct() -> panc_v_ctrl_features


toil.quant$norm.counts.tidy %>% 
  # filter(gene %in% patient.plasma.quant$de.seq$gencode.34.salmon.out$gene) %>% 
  filter(gene %in% filter(salmon.quant$panc_v_ctrl$deseq$de, log2FoldChange <= -1)$gene) %>%
  select(gene, sample, count, type) %>%
  # filter(gene %in% panc_v_ctrl_features$gene) %>% 
  filter(type %in% c('Primary Tumor', 'Normal Tissue')) %>% 
  mutate(type = as.factor(ifelse(type == 'Normal Tissue', 0, 1))) %>% 
  # spread(gene, count) %>% 
  pivot_wider(names_from = gene, values_from = count, values_fn = mean) %>% 
  column_to_rownames('sample') %>% 
  initial_split(prop = .75, strata = 'type') -> toil.init.split

toil.quant$norm.counts.tidy %>% 
  # filter(gene %in% patient.plasma.quant$de.seq$gencode.34.salmon.out$gene) %>% 
  # filter(gene %in% filter(salmon.quant$panc_v_ctrl$deseq$de, log2FoldChange <= -1)$gene) %>% 
  select(gene, sample, count, type) %>%
  filter(type %in% c('Primary Tumor', 'Normal Tissue'),
         gene %in% colnames(toil.small.n.train)) %>% 
  mutate(type = as.factor(ifelse(type == 'Normal Tissue', 0, 1))) %>% 
  # spread(gene, count) %>% 
  pivot_wider(names_from = gene, values_from = count, values_fn = mean) %>% 
  column_to_rownames('sample') -> training.data

  
toil.train.split <- training(toil.init.split)
toil.test.split <- testing(toil.init.split)

salmon.quant$panc_v_ctrl$deseq$counts %>%
  # filter(gene %in% panc_v_ctrl_features$gene) %>%
  select(-ensg) %>% 
  # filter(gene %in% colnames(toil.train.split)) %>%
  # rownames_to_column('gene') %>%
  # filter_at(vars(contains('.')), all_vars(. >= 1)) %>%
  filter(gene %in% filter(salmon.quant$panc_v_ctrl$deseq$de, log2FoldChange <= -1)$gene) %>%
  gather(patient, count, -gene) %>%
  merge(current.meta.data %>% select(patient, condition) %>% filter(condition %in% c('ctrl', 'panc'))) %>%
  distinct() %>%
  mutate(type = as.factor(ifelse(condition == 'ctrl', 0, 1))) %>%
  select(patient, gene, count, type) %>%
  # merge(gencode.35.tx.to.gene %>% select(gene, ensg), by = 'ensg') %>%
  distinct() %>%
  # select(gene, count, patient, type) %>%
  pivot_wider(names_from = gene, values_from = count, values_fn = mean) %>% 
  column_to_rownames('patient') -> toil.small.n.train

norm_model <- train(
  type ~ .,
  data = toil.train.split,
  # data = toil.small.n.train,
  method = 'glm',
  family = 'binomial',
  preProc = c('zv', 'center', 'scale'),
  trControl = trainControl(method = 'cv', number = 10)
)

#modelLookup('glm')

penalized_model <- train(
  type ~ .,
  data = toil.train.split,
  # data = toil.small.n.train,
  method = 'glmnet',
  family = 'binomial',
  preProc = c('zv', 'center', 'scale'),
  trControl = trainControl(method = 'cv', number = 10),
  tuneLength = 10
)

penalized_curve <- caret::learning_curve_dat(
  dat = toil.train.split,
  # dat = toil.small.n.train,
  outcome = 'type',
  # proportion = 1,
  test_prop = 1/4,
  method = 'glmnet',
  family = 'binomial',
  preProc = c('zv', 'center', 'scale'),
  # trControl = trainControl(method = 'cv', number = 10),
  # tuneLength = 10
)


vip(norm_model)

vip::vi(penalized_model)
vip(penalized_model)

ggplot(penalized_curve, aes(x = Training_Size, y = Accuracy, color = Data)) + 
  geom_smooth(method = loess, span = .8)

summary(resamples(list(reg = norm_model,
                      pen = penalized_model)))$statistics$Accuracy
#####
norm_pred <- predict(norm_model, toil.test.split)

confusionMatrix(data = relevel(norm_pred, ref = '1'),
                reference = relevel(toil.test.split$type, ref = '1'))

penalized_pred <- predict(penalized_model, toil.test.split)

confusionMatrix(data = relevel(penalized_pred, ref = '1'),
                reference = relevel(toil.test.split$type, ref = '1'))
#####
norm_pred <- predict(norm_model, training.data)

confusionMatrix(data = relevel(norm_pred, ref = '1'),
                reference = relevel(training.data$type, ref = '1'))

penalized_pred <- predict(penalized_model, training.data)

confusionMatrix(data = relevel(penalized_pred, ref = '1'),
                reference = relevel(training.data$type, ref = '1'))

#####
norm_validation <- predict(norm_model, toil.small.n.train)

confusionMatrix(data = relevel(norm_validation, ref = '1'),
                reference = relevel(toil.small.n.train$type, ref = '1'))

penalized_validation <- predict(penalized_model, toil.small.n.train)

confusionMatrix(data = relevel(penalized_validation, ref = '1'),
                reference = relevel(toil.small.n.train$type, ref = '1'))
#####
predict(penalized_model, toil.small.n.train, type = 'prob') %>% 
  select(`1`) %>%
  prediction(., toil.small.n.train$type) %>% 
  performance(measure = 'tpr', x.measure = 'fpr') -> penalized_performance

penalized_performance_df <- data_frame('fpr' = unlist(penalized_performance@x.values), 
                                       'tpr' = unlist(penalized_performance@y.values))

penalized_performance_df %>% 
  ggplot(aes(fpr, tpr)) + 
    geom_line()


merge(patient.plasma.quant$de.seq$gencode.34.salmon.out, panc1.intra.quant$de.seq$gencode.34.salmon.out, by = 'gene') %>% 
  filter(gene %in% patient.plasma.quant$rfe.condition.profile$optVariables) %>% 
  ggplot(aes(log2FoldChange.x, log2FoldChange.y, label = gene)) + 
  geom_point() + 
  geom_text()

# norm_pred <- predict(norm_model, toil.total.test)
# 
# confusionMatrix(data = relevel(norm_pred, ref = '1'),
#                 reference = relevel(toil.total.test$type, ref = '1'))
# 
# penalized_pred <- predict(penalized_model, toil.total.test)

# confusionMatrix(data = relevel(penalized_pred, ref = '1'),
#                 reference = relevel(toil.total.test$type, ref = '1'))


# toil.quant$norm.counts.tidy.z.scores %>% 
#   # filter(gene %in% patient.plasma.quant$rfe.kras.sig.condition.profile$optVariables) %>%
#   filter(gene %in% filter(panc1.intra.quant$de.seq$gencode.34.salmon.out, log2FoldChange < -1)$gene) %>%
#   # filter(gene %in% patient.plasma.quant$rfe.lnc.condition.profile$optVariables) %>%
#   # filter(gene %in% gen.34.lnc.names$gene, 
#   #        gene %in% filter(patient.plasma.quant$de.seq$gencode.34.salmon.out, log2FoldChange > 1)$gene) %>% 
#   filter(gene %in% filter(patient.plasma.quant$de.seq$gencode.34.salmon.out, log2FoldChange > 1)$gene)  %>%
#   select(gene, sample, count, type) %>%
#   filter(type %in% c('Primary Tumor', 'Normal Tissue')) %>% 
#   mutate(type = as.factor(ifelse(type == 'Normal Tissue', 0, 1))) %>% 
#   spread(gene, count) %>% 
#   column_to_rownames('sample') %>% 
#   initial_split(prop = .6, strata = 'type') -> toil.init.split
#   
# toil.train.split <- training(toil.init.split)
# toil.test.split <- testing(toil.init.split)  
# 
# norm_model <- train(
#   type ~ .,
#   data = toil.train.split,
#   method = 'glm',
#   family = 'binomial',
#   preProc = c('zv', 'center', 'scale'),
#   trControl = trainControl(method = 'cv', number = 10)
# )
# 
# penalized_model <- train(
#   type ~ .,
#   data = toil.train.split,
#   method = 'glmnet',
#   family = 'binomial',
#   preProc = c('zv', 'center', 'scale'),
#   trControl = trainControl(method = 'cv', number = 10),
#   tuneLength = 10
# )
# 
# summary(resamples(list(reg = norm_model,
#                       pen = penalized_model)))$statistics$Accuracy
# 
# vip(norm_model, all_permutations = T, bar = T)
# vip(penalized_model, all_permutations = T, bar = T, jitter = T)
# 
# norm_pred <- predict(norm_model, toil.test.split)
# 
# confusionMatrix(data = relevel(norm_pred, ref = '1'),
#                 reference = relevel(toil.test.split$type, ref = '1'))
# 
# penalized_pred <- predict(penalized_model, toil.test.split)
# 
# confusionMatrix(data = relevel(penalized_pred, ref = '1'),
#                 reference = relevel(toil.test.split$type, ref = '1'))


```
### A: Biotype distribution 
```{r}
biotype.dist <- function(tx.counts){
  return(tx.counts %>% 
    rownames_to_column('enst') %>% 
    merge(gencode.35.tx.to.gene, by = 'enst') %>% 
    gather(sample, count, -enst, -ensg:-biotype) %>% 
    mutate(sample.parse = sample) %>% 
    separate(sample.parse, sep = '[.]', 
           into = c('platform', 'context', 'rep', 'condition')) %>% 
    group_by(biotype, sample, condition) %>% 
    summarise(count = sum(count)) %>% 
    group_by(sample) %>% 
    mutate(freq = count / sum(count)) %>% 
    mutate(biotype = ifelse(biotype %in% c('lncRNA', 'Mt_rRNA', 'protein_coding',
                                           'non_stop_decay', 'nonsense_mediated_decay', 'retained_intron',
                                           'processed_transcript', 'Alu', 'L1', 'ERV1'),
                            biotype, 'other'))) #%>%
    # mutate(biotype = ifelse(biotype %in% te.families.clades$family, 'TE', biotype)))
}

tot.biotype.dist <- function(tx.counts){
  return(tx.counts %>% 
    rownames_to_column('enst') %>% 
    merge(total.tx.to.gene, by = 'enst') %>% 
    gather(sample, count, -enst, -ensg:-biotype) %>% 
    mutate(sample.parse = sample) %>% 
    separate(sample.parse, sep = '[.]', 
           into = c('platform', 'context', 'rep', 'condition')) %>% 
    group_by(biotype, sample, condition) %>% 
    summarise(count = sum(count)) %>% 
    group_by(sample) %>% 
    mutate(freq = count / sum(count)) %>% 
    mutate(biotype = ifelse(biotype %in% c('lncRNA', 'Mt_rRNA', 'protein_coding', 'retained_intron',
                                           'Alu', 'L1', 'ERV1'),
                            biotype, 'other'))) #%>%
    # mutate(biotype = ifelse(biotype %in% te.families.clades$family, 'TE', biotype)))
}


biotype.dist.plt <- function(tx.biotype.dist){
  tx.biotype.dist %>% 
    merge(metadata.df, by.x = 'sample', by.y = 'names') %>% 
    distinct() %>% 
    unite(condition, c(condition.x, context)) %>% 
    mutate(condition = factor(condition, levels = c('DMSO_2D', 'AMG_2D', 'DMSO_3D', 'AMG_3D')),
           tmp = sample) %>% 
    separate(tmp, sep='[.]', into = c(NA, NA, 'rep', NA)) %>% 
    ggplot(aes(rep, freq, fill = biotype, color = biotype)) +
      geom_bar(stat = 'identity', color = 'white') +
      scale_fill_viridis_d( direction = 1) +
      # scale_color_viridis_d(direction = 1) +
      scale_x_discrete(expand = c(0, 0.5)) +
      # scale_y_continuous(expand = c(0.5,0.5)) + 
      # geom_hline(yintercept = seq(0,1,0.25), color = 'white') +
      xlab('') + ylab('% of total counts') +
      # theme(axis.text.x = element_text(angle = 40, hjust = 1)) + 
    facet_row(~condition, scales = 'free_x')
}


tot.biotype.dist.plt <- function(tx.counts, control = 'ctrl', treatment){
  tx.counts %>% 
    rownames_to_column('enst') %>% 
    merge(total.tx.to.gene, by = 'enst') %>% 
    gather(sample, count, -enst, -ensg:-biotype) %>% 
    merge(current.meta.data, by.x = 'sample', by.y = 'patient') %>% 
    # filter(biotype != 'Mt_rRNA') %>% 
    group_by(biotype, sample, condition) %>% 
    summarise(count = sum(count)) %>% 
    group_by(sample) %>% 
    mutate(freq = count / sum(count)) #%>% 
    # mutate(biotype = ifelse(biotype %in% c('lncRNA', 'protein_coding', 'Alu', 'L1', 'ERV1'),
    #                         biotype, 'other'),
    #        condition = ifelse(grepl('ctrl', sample), control, treatment)) %>% 
    # mutate(condition = ifelse(grepl('ctrl', sample), control, treatment)) %>% 
    # mutate(biotype = ifelse(biotype %in% te.families.clades$family, 'TE', biotype)) %>% 
    # ggplot(aes(sample, freq, fill = biotype, color = biotype)) + 
    #   geom_bar(stat = 'identity') +
    #   # geom_boxplot(outlier.size = NULL,position = position_dodge(width = 1)) + 
    #   #geom_sina(position = position_dodge(width = 1)) +
    #   scale_fill_viridis_d( direction = -1) +
    #   scale_color_viridis_d(direction = -1) + 
    #   scale_x_discrete(expand = c(0, 0.5)) +
    #   geom_hline(yintercept = seq(0,1,0.25), color = 'white') + 
    #   xlab('') + ylab('% of total counts') + 
    #   theme(axis.text.x = element_text(angle = 40, hjust = 1))
    # ggplot(aes(reorder(biotype, -freq), freq, fill = condition, color = condition)) +
    #   geom_boxplot() +
    #   # geom_sina() +
    #   scale_fill_viridis_d( direction = -1) +
    #   scale_color_viridis_d(direction = -1) +
    #   theme(axis.text.x = element_text(angle = 40, hjust = 1))
}
```
####
```{r}

salmon.quant$rnaEasy_all$deseq$biotype.dist <- biotype.dist(salmon.quant$rnaEasy_all$deseq$txi.counts)

biotype.dist.plt(salmon.quant$rnaEasy_all$deseq$biotype.dist)


ucsc.rmsk.salmon.quant$rnaEasy_all$deseq$biotype.dist <-
  tot.biotype.dist(ucsc.rmsk.salmon.quant$rnaEasy_all$deseq$txi.counts)

biotype.dist.plt(ucsc.rmsk.salmon.quant$rnaEasy_all$deseq$biotype.dist)

merge(biotype.dist, tot.biotype.dist, by = c('biotype', 'sample')) %>% 
  select(biotype, sample, condition = condition.x, gencode_freq = freq.x, total_freq = freq.y) %>% 
  # filter_at(vars(contains('_freq')), all_vars(. >= 0.01)) %>% 
  filter(biotype %in% c('protein_coding', 'Mt_rRNA', 'lncRNA', 'processed_transcript', 'nonsene_mediated_decay', 'retained_intron')) %>% 
  mutate(freq_diff = total_freq - gencode_freq) %>% 
  ggplot(aes(reorder(biotype, freq_diff), freq_diff, color = condition)) + 
  geom_boxplot(position = position_dodge(width = 1)) + 
  geom_sina(position = position_dodge(width = 1)) + 
  scale_color_viridis_d()
  

```

### Tx coverage
```{r}
salmon.quant$rnaEasy_all$deseq$tx.cov <- tx.ome.cov.plt(salmon.quant$rnaEasy_all$deseq$txi.counts)


salmon.quant$rnaEasy_all$deseq$tx.cov %>% 
  separate(names, sep = '[.]', into = c('platform', 'context', 'rep', 'condition')) %>% 
  mutate(condition = factor(condition, levels = c('DMSO', 'AMG'))) %>% 
  unite(col = 'condition', c('condition', 'context')) %>% 
  mutate(condition = factor(condition, levels = c('DMSO_2D', 'AMG_2D', 'DMSO_3D', 'AMG_3D'))) %>% 
  group_by(condition, platform, rep) %>% 
  summarize(tot.tx.count = sum(biotype.count)) %>% 
  ggplot(aes(condition, tot.tx.count)) + 
  geom_boxplot(width = .3, position = position_dodge(width = 1), outlier.color = NA, alpha = 0.3) +
  geom_sina(width = .3, alpha = 0.5, position = position_dodge(width = 1)) +
  # ylim(6000, 16000) +
  # scale_color_viridis_d() +
  ylab('transcripts detected') + xlab('') + theme(legend.position = c(1,0.95)) +
  stat_compare_means(comparisons = list(c('DMSO_2D', 'AMG_2D'), c('DMSO_2D', 'DMSO_3D'),
                                        c('DMSO_3D', 'AMG_3D'), c('AMG_2D', 'AMG_3D')), method = 'wilcox.test')

```

# Figures



## Fig 1.

### A: Volcano
```{r}
gencode.de.volcano <- function(de.seq){
  de.seq <- 
    de.seq %>% filter(!grepl('^MT', gene)) %>% 
    mutate(biotype = ifelse(gene %in% gencode.35.lnc.gene.names$gene, 'lncRNA','coding'),
           padj = ifelse(padj == 0, min(filter(de.seq, padj != 0)$padj), padj))
  x.max <- round(max(abs(de.seq$log2FoldChange)) + 0.5)
  print(x.max)
  label.max <- 0.01
  ggplot(de.seq, 
         aes(log2FoldChange, -log10(padj), 
             color = biotype)) +
      geom_point(alpha = 0.75, size = rel(2)) + 
      geom_rug(size = rel(1), alpha = 0.15) + 
      scale_x_continuous(limits = c(-x.max,x.max), expand = c(0,0.5),
                         breaks = c(-x.max, -x.max/2, -1, 0, 1, x.max/2, x.max)) +
      scale_color_manual(values = c(viridis(1)[[1]], viridis(3)[[2]], viridis(5)[[5]])) +
      xlab('log2(Fold Change)') +
      ylab('-log10(p.adj)') + 
      geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) +
      # geom_text_repel(aes(label = ifelse(padj <= label.max & log2FoldChange <= -x.max/1.5| log2FoldChange >= x.max/1.5,
      #                                    as.character(gene), '')),
      geom_text_repel(aes(label = as.character(gene)),
                              box.padding = unit('1', 'lines'),
                              segment.color = 'grey',
                              segment.alpha = 0.2,
                              color = 'black',
                              ylim = c(1, Inf), size = base_size/2, family = 'Helvetica')
}

fgsea.hallmark.de.volcano <- function(fgsea){
  fgsea <- 
    fgsea %>% filter(grepl('HALLMARK|SWEET', pathway))
  x.max <- round(max(abs(fgsea$NES)) + 0.5)
  ggplot(fgsea, 
         aes(NES, -log10(padj), 
             color = size, 
             label = tolower(sub('HALLMARK', '', pathway)))) +
      geom_point(alpha = 0.6) + 
      geom_rug(size = rel(1), alpha = 0.6) + 
      scale_size_continuous(name = 'genes') +
      scale_x_continuous(limits = c(-x.max,x.max), expand = c(0,0.5),
                         breaks = c(-x.max, -x.max/2, -1, 0, 1, x.max/2, x.max)) +
      scale_color_viridis_c() +
      xlab('NES') +
      ylab('-log10(p.adj)') + 
      geom_text_repel(box.padding = unit('0.8', 'lines'),
                              segment.color = 'grey',
                              segment.alpha = 0.2,
                              color = 'black',
                              ylim = c(2.0, Inf), size = rel(1.5))
}
```
####
```{r}

gencode.de.volcano(salmon.quant$exoTIC_2D$deseq$de) + theme(legend.position = c(0.95, 0.95))
gencode.de.volcano(salmon.quant$exoTIC_3D$deseq$de) + theme(legend.position = c(0.15, 0.95))

gencode.de.volcano(aale.salmon.quant$deseq$de %>% 
                     filter(gene %in% g12c.marker.genes$induced)) + 
  theme(legend.position = c(0.125, 0.95))

gencode.de.volcano(salmon.quant$exoTIC_2D$deseq$de %>% filter(grepl('ZNF', gene))) #+ 
  theme(legend.position = c(0.95, 0.95))

gencode.de.volcano(salmon.quant$exoTIC_2D$deseq$de %>% 
                     filter(gene %in% g12c.sc.marker.genes$`G12C supressed`)) + 
  theme(legend.position = c(0.125, 0.95))

gencode.de.volcano(salmon.quant$rnaEasy_all$deseq$de) + theme(legend.position = c(0.95, 0.95))
# gencode.35.tx.to.gene %>% 
#   filter(gene %in% filter(salmon.quant$rnaEasy_2D$deseq$de, gene %in% gencode.35.lnc.gene.names$gene)$gene) %>%
#   filter(gene %in% filter(salmon.quant$rnaEasy_2D$deseq$de, log2FoldChange >= 1)$gene) %>%
#   merge(as.data.frame(rowRanges(salmon.quant$gxi)), by.x = 'ensg', by.y = 'gene_id') %>% 
#   select(seqnames, start, end, strand) %>% 
#   distinct() %>% 
#   print(row.names = F)
# 
# gencode.35.tx.to.gene %>% 
#   filter(gene %in% gencode.35.lnc.gene.names$gene) %>%
#   # filter(!gene %in% filter(salmon.quant$rnaEasy_3D$deseq$de, log2FoldChange >= 1)$gene) %>%
#   merge(as.data.frame(rowRanges(salmon.quant$gxi)), by.x = 'ensg', by.y = 'gene_id') %>% 
#   select(seqnames, start, end, strand) %>% 
#   distinct() -> lnc.bg.bed
# 
# write_tsv(lnc.bg.bed, file = file.path(output.data.dir, 'lnc.bg.bed'), col_names = F)

salmon.quant$three_D$deseq$de %>% 
  merge(salmon.quant$two_D$deseq$de, by = 'gene') %>% 
  ggplot(aes(log2FoldChange.x, log2FoldChange.y)) + 
    geom_point() +
    ylab('rnaEasy') + xlab('exoTIC') + 
    geom_hline(yintercept = 0, linetype = 'dashed', alpha=0.2) + 
    geom_vline(xintercept = 0, linetype = 'dashed', alpha=0.2)

salmon.quant$exoTIC_2D$deseq$de %>% 
  # filter(gene %in% g12c.marker.genes$suppressed) %>% 
  merge(aale.salmon.quant$deseq$de, by = 'gene') %>% 
  mutate(biotype = ifelse(gene %in% gencode.35.lnc.gene.names$gene, 'lncRNA','coding')) %>% 
  filter(!grepl('^RPS|^MT', gene)) %>% 
  ggplot(aes(log2FoldChange.x, log2FoldChange.y, color = biotype)) + 
    geom_point() +
    xlab('exoTIC_2D') + ylab('aale_Exo') + 
    geom_hline(yintercept = 0, linetype = 'dashed', alpha=0.2) + 
    geom_hline(yintercept = c(-1,1), linetype = 'dashed', alpha=0.1) + 
    geom_vline(xintercept = 0, linetype = 'dashed', alpha=0.2) +
    geom_vline(xintercept = c(-1,1), linetype = 'dashed', alpha=0.1) + 
    scale_color_manual(values = c(viridis(1)[[1]], viridis(3)[[2]], viridis(5)[[5]])) +
    # scale_x_continuous(limits = c(-4, 4), breaks = c(-4, -2, -1, 0, 1, 2, 4)) +
    geom_text_repel(aes(label = ifelse(abs(log2FoldChange.x) >= 2 | abs(log2FoldChange.y) >= 2,
                                         as.character(gene), '')),
                              box.padding = unit('1', 'lines'),
                              segment.color = 'grey',
                              segment.alpha = 0.2,
                              color = 'black',
                              ylim = c(-Inf, Inf), size = base_size/3, family = 'Helvetica') +
    theme(legend.position = c(1, 0.85))


```
### B: Tx-ome coverage
```{r}
tx.ome.cov.plt <- function(tx.counts, ctrl = 'ctrl', treatment){
  tx.counts %>% 
    rownames_to_column('enst') %>% 
    merge(gencode.35.tx.to.gene, by = 'enst') %>% 
    gather(sample, count, -enst, -ensg:-biotype) %>% 
    separate(sample, sep = '[.]', 
           into = c('platform', 'context', 'rep', 'condition')) %>% 
    # rownames_to_column()
    # gather(sample, count, -tx:-biotype) %>% 
    # merge(current.meta.data, by.x = 'sample', by.y = 'patient') %>% 
    # filter(biotype != 'Mt_rRNA') %>% 
    # mutate(condition = ifelse(grepl(ctrl, sample), ctrl, treatment)) %>% 
    filter(count >= 5) %>% 
    group_by(biotype, sample, condition) %>% 
    mutate(biotype.count = n()) %>% 
    select(biotype, sample, condition, biotype.count) %>% 
    distinct() %>% 
    merge(gen.35.biotype.count, by = 'biotype') %>% 
    mutate(coverage = biotype.count/n) %>% 
    # filter(biotype %in% c('lncRNA', 'protein_coding', 'retained_intron', 'pseudogene')) %>% 
    filter(biotype %in% c('protein_coding', 'lncRNA', 'processed_transcript', 'nonsene_mediated_decay', 'retained_intron')) %>% 
    ggplot(aes(reorder(biotype, -coverage), coverage, color = condition)) + 
    geom_boxplot(position = position_dodge(width = 1)) +
    geom_sina(position = position_dodge(width = 1)) +
    # scale_color_manual(values = c(viridis(1)[[1]], viridis(3)[[2]])) +
    scale_color_viridis_d() + 
    xlab('') + ylab('biotype complexity') + 
    theme(axis.text.x = element_text(angle = 40, hjust = 1))
}

tot.biotype.cov.plt <- function(tx.counts, control = 'ctrl', treatment){
  tx.counts %>% 
    rownames_to_column('enst') %>% 
    merge(total.tx.to.gene, by = 'enst') %>% 
    gather(sample, count, -enst, -ensg:-biotype) %>% 
    # merge(current.meta.data, by.x = 'sample', by.y = 'patient') %>% 
    # rownames_to_column()
    # gather(sample, count, -tx:-biotype) %>% 
    # merge(current.meta.data, by.x = 'sample', by.y = 'patient') %>% 
    filter(biotype != 'Mt_rRNA') %>% 
    # mutate(condition = ifelse(grepl(ctrl, sample), ctrl, treatment)) %>% 
    filter(count >= 5) %>% 
    group_by(biotype, sample, condition) %>% 
    mutate(biotype.count = n()) %>% 
    select(biotype, sample, condition, biotype.count) %>% 
    distinct() %>% 
    merge(total.tx.to.gene.count %>% filter(biotype %in% c('Alu', 'ERV1', 'L1', 'L2',
                                                           'ERVK', 'ERVL', 'ERV-MaLR',
                                                           'protein_coding', 'lncRNA', 'processed_transcript', 
                                                           'nonsene_mediated_decay', 'retained_intron')), by = 'biotype') %>% 
    mutate(coverage = biotype.count/n) %>% 
    # filter(biotype %in% c('lncRNA', 'protein_coding', 'retained_intron', 'pseudogene')) %>% 
    # filter(biotype %in% c('protein_coding', 'lncRNA', 
    #                       'processed_transcript', 'nonsene_mediated_decay', 
    #                       'retained_intron', 'Alu', 'ERV1', 'L1')) %>% 
    ggplot(aes(reorder(biotype, -coverage), coverage, color = condition)) + 
    geom_boxplot(position = position_dodge(width = 1)) +
    geom_sina(position = position_dodge(width = 1)) +
    # scale_color_manual(values = c(viridis(1)[[1]], viridis(3)[[2]])) +
    scale_color_viridis_d() + 
    xlab('') + ylab('biotype complexity') + 
    theme(axis.text.x = element_text(angle = 40, hjust = 1))
}

tx.ome.cov.plt <- function(tx.counts, ctrl = 'ctrl', treatment){
  tx.counts %>% 
    rownames_to_column('enst') %>% 
    merge(gencode.35.tx.to.gene, by = 'enst') %>% 
    gather(sample, count, -enst, -ensg:-biotype) %>% 
    mutate(condition= sample) %>% 
    separate(condition, sep = '[.]', into = c('platform', 'context', 'rep', 'condition')) %>% 
    filter(biotype != 'Mt_rRNA') %>% 
    filter(count >= 5) %>% 
    group_by(biotype, sample, condition) %>% 
    mutate(biotype.count = n()) %>% 
    select(biotype, sample, condition, biotype.count) %>% 
    distinct() %>% 
    merge(gen.35.biotype.count, by = 'biotype') %>% 
    mutate(coverage = biotype.count/n) 
}

tot.tx.ome.cov.plt <- function(tx.counts, ctrl = 'ctrl', treatment){
  tx.counts %>% 
    rownames_to_column('enst') %>% 
    merge(total.tx.to.gene %>% 
            filter(biotype %in% c('Alu', 'ERV1', 'L1', 'L2',
                                  'ERVK', 'ERVL', 'ERV-MaLR',
                                  'protein_coding', 'lncRNA', 'processed_transcript',
                                  'nonsene_mediated_decay', 'retained_intron')), by = 'enst') %>% 
    gather(sample, count, -enst, -ensg:-biotype) %>% 
    mutate(condition= sample) %>% 
    separate(condition, sep = '[.]', into = c('platform', 'context', 'rep', 'condition')) %>% 
    filter(biotype != 'Mt_rRNA') %>% 
    filter(count >= 5) %>% 
    group_by(biotype, sample, condition) %>% 
    mutate(biotype.count = n()) %>% 
    select(biotype, sample, condition, biotype.count) %>% 
    distinct() %>% 
    merge(gen.35.biotype.count, by = 'biotype') %>% 
    mutate(coverage = biotype.count/n) 
}
```
####
```{r}
# tx.ome.cov.plt(patient.plasma.quant$tx$gencode.34.salmon.out, treatment = 'panc')

txome.cov.dist <- tx.ome.cov.plt(salmon.quant$deseq$txi.counts, treatment = 'panc')

txome.cov.dist %>% 
  mutate(context = sample) %>% 
  filter(biotype == 'protein_coding') %>% 
  separate(context, sep = '[.]', 
           into = c('platform', 'context', 'rep', NA)) %>% 
  mutate(condition = factor(condition, levels = c('DMSO', 'AMG'))) %>% 
  group_by(sample, condition, platform, context) %>% 
  summarize(tot.tx.count = sum(biotype.count)) %>% 
  ggplot(aes(context, tot.tx.count, color = condition)) + 
  geom_boxplot(width = .3, position = position_dodge(width = 1), outlier.color = NA, alpha = 0.3) +
  geom_sina(position = position_dodge(width = 1), alpha = 0.5) +
  scale_color_manual(values = c(viridis(20)[16], viridis(20)[1])) +
  ylab('transcripts detected') + xlab('') + facet_row(~platform)
  # stat_compare_means(comparisons = list(c('ctrl', 'panc'), c('ctrl', 'covid'), c('panc', 'covid')), 
  #                    method = 'wilcox.test')

tot.txome.cov.dist <- tot.tx.ome.cov.plt(ucsc.rmsk.salmon.quant$deseq$txi.counts, treatment = 'panc')


tot.txome.cov.dist %>% 
  mutate(context = sample) %>% 
  # filter(biotype == 'protein_coding') %>% 
  separate(context, sep = '[.]', 
           into = c('platform', 'context', 'rep', NA)) %>% 
  mutate(condition = factor(condition, levels = c('DMSO', 'AMG'))) %>% 
  group_by(sample, condition, platform, context) %>% 
  summarize(tot.tx.count = sum(biotype.count)) %>% 
  ggplot(aes(context, tot.tx.count, color = condition)) + 
  geom_boxplot(width = .3, position = position_dodge(width = 1), outlier.color = NA, alpha = 0.3) +
  geom_sina(position = position_dodge(width = 1), alpha = 0.5) +
  scale_color_manual(values = c(viridis(20)[16], viridis(20)[1])) +
  ylab('transcripts detected') + xlab('') + facet_row(~platform)


tot.biotype.cov.plt(ucsc.rmsk.salmon.quant$deseq$txi.counts, treatment = 'panc')

# /
# tx.ome.cov.plt(panc1.intra.quant$tx$gencode.34.salmon.out, treatment = 'kras_kd')/
# tx.ome.cov.plt(panc1.exo.intra.compare$tx$gencode.34.salmon.out, 'intra', 'exo')

# patient.plasma.quant$tx$gencode.34.salmon.out %>% 
#   gather(sample, count, -tx:-biotype) %>% 
#   filter(biotype != 'Mt_rRNA') %>% 
#   mutate(condition = ifelse(grepl('ctrl', sample), 'ctrl', 'panc')) %>% 
#   filter(count >= 10) %>% 
#   group_by(biotype, sample, condition) %>% 
#   mutate(biotype.count = n()) %>% 
#   select(biotype, sample, condition, biotype.count) %>% 
#   distinct() %>% 
#   merge(gen.34.tx.2.gene.biotype.count, by = 'biotype') %>% 
#   mutate(coverage = biotype.count/n) %>% 
#   filter(biotype %in% c('lncRNA', 'protein_coding', 'retained_intron', 'pseudogene')) %>% 
#   ggplot(aes(reorder(biotype, -coverage), coverage*100, color = condition)) + 
#   geom_boxplot() +
#   scale_color_viridis_d(direction = -1) + 
#   xlab('') + ylab('% tx coverage') + 
#   theme(axis.text.x = element_text(angle = 40, hjust = 1))
```

### C: rerwip Tissue coverage
```{r}
## not sure if norm is working well
tissue.cov.plt <- function(gene.counts){
  gene.counts %>% 
    # merge(gencode.35.tx.to.gene %>% select('ensg', 'gene'), by = 'ensg') %>% 
    select(-ensg) %>% 
    gather(sample, count, -gene) %>% 
    # mutate(condition = ifelse(grepl('ctrl', sample), 'ctrl', treatment)) %>%
    separate(sample, sep = '[.]', 
           into = c('platform', 'context', 'rep', 'condition')) %>% 
    # merge(current.meta.data, by.x = 'sample', by.y = 'patient') %>% 
    merge(yang.ts.score.gtex, 
          by = 'gene') %>% 
    group_by(platform, context, condition) %>% 
    mutate(sample.counts = sum(count)) %>% 
    group_by(tissue) %>% 
    mutate(tissue.counts = n()) %>% 
    group_by(platform, context, condition, tissue) %>% 
    mutate(tissue.calc = ((count/sample.counts)/tissue.counts)) %>% 
    group_by(platform, context, condition, tissue) %>% 
    summarise(sum.tissue.calc = sum(tissue.calc)) %>% 
    ggplot(aes(reorder(tissue, -log(sum.tissue.calc)),log(sum.tissue.calc), 
               fill = condition)) + 
    geom_boxplot(position = position_dodge(width = 1)) + 
    geom_sina(size = rel(1), alpha = 0.6, position = position_dodge(width = 1)) +
    scale_fill_viridis_d(alpha = 0.6, direction = -1) +
    scale_color_viridis_d(direction = -1) + 
    xlab('') + ylab('norm tissue score') + 
    theme(axis.text.x = element_text(angle = 40, hjust = 1))
}
```
####
```{r}

tissue.cov.plt(salmon.quant$exoTIC_2D$deseq$counts)

```

### D: Hclust
```{r}
gene.expression.hclust.plt <- function(gene.counts){
  gene.counts %>% 
    select(-ensg, gene, contains('.')) %>% 
    # rownames_to_column('gene') %>% 
    # filter(gene %in% c('KRAS', 'NRAS', 'MRAS', 'RRAS', 'HRAS')) %>%
    filter(gene %in% g12c.marker.genes$induced) %>%
    # filter(gene %in% salmon.quant$rnaEasy_2D$deseq$de$gene) %>% 
    filter_at(vars(contains('.')), any_vars(. >= 10)) %>%
    distinct() %>% 
    gather(sample, count, -gene) %>% 
    # mutate(sample.parse = sample) %>% 
    # separate(sample.parse, sep = '[.]', 
    #        into = c('platform', 'context', 'rep', 'condition')) %>%
    # merge(merged.patient.metadata %>% select(patient, batch), by.x = 'patient', by.y = 'patient') %>% 
    filter(!grepl('^MT', gene)) %>% 
    mutate(count = log(count + 1),
           count = scale(count, center = T)) %>% 
    spread(sample, count) %>% 
    column_to_rownames('gene') %>%
    pheatmap(treeheight_row = 10, treeheight_col = 10, show_rownames = F, color = viridis(32, option = 'magma'))
}


tx.expression.hclust.plt <- function(tx.counts){
  tx.counts %>% 
    filter_at(vars(contains('.')), all_vars(. >= 1)) %>% 
    gather(patient, count, -tx:-biotype) %>% 
    filter(!grepl('^MT', gene)) %>% 
    mutate(count = log(count),
           count = scale(count, center = T)) %>% 
    spread(patient, count) %>% 
    column_to_rownames('tx') %>% 
    select(-gene, -len, -biotype) %>% 
    pheatmap(treeheight_row = 10, treeheight_col = 10,
             show_rownames = F, color = viridis(32), 
             border_color = NA)
}

```
####
```{r}
gene.expression.hclust.plt(salmon.quant$rnaEasy_all$deseq$counts)


salmon.quant$rnaEasy_all$deseq$counts %>% 
  select(-ensg, gene, contains('.')) %>% 
  filter(gene %in% g12c.marker.genes$induced) %>% 
  gather(sample, count, -gene) %>% 
  mutate(count = log(count + 1),
         count = scale(count, center = T)) %>% 
  pivot_wider(names_from = sample, values_from = count, values_fn = mean) %>% 
  column_to_rownames('gene') %>%
  pheatmap(treeheight_row = 10, treeheight_col = 10, show_rownames = F)


salmon.quant$rnaEasy_all$deseq$counts %>% 
  # separate(gene, sep = '_', into = c('gene', 'ensg')) %>% 
  filter(gene %in% g12c.marker.genes$induced) %>%
  unite(col = 'gene', sep = '_', c('gene', 'ensg')) %>%
  filter_at(vars(contains('.')), any_vars(. >= 1)) %>%
  column_to_rownames('gene') %>% 
  as.matrix() %>% cor() %>% pheatmap(color = viridis(32, option = 'magma'))


salmon.quant$deseq$combat.counts %>% 
  # separate(gene, sep = '_', into = c('gene', 'ensg')) %>% 
  # filter(gene %in% gencode.35.lnc.gene.names$gene) %>%
  # unite(col = 'gene', sep = '_', c('gene', 'ensg')) %>%
  filter_at(vars(contains('.')), any_vars(. >= 1)) %>%
  column_to_rownames('gene') %>% 
  as.matrix() %>% cor() -> batch.mtx

salmon.quant$deseq$counts %>% 
  # separate(gene, sep = '_', into = c('gene', 'ensg')) %>% 
  # filter(gene %in% gencode.35.lnc.gene.names$gene) %>%
  unite(col = 'gene', sep = '_', c('gene', 'ensg')) %>%
  filter_at(vars(contains('.')), any_vars(. >= 1)) %>%
  column_to_rownames('gene') %>% 
  as.matrix() %>% cor() -> norm.mtx

norm.mtx[1,2] - batch.mtx[1,2] 

norm.mtx - batch.mtx -> diff.mtx

diff.mtx %>% pheatmap(color = viridis(32, option = 'magma'), 
                                     annotation_col = annotation.row.df)


salmon.quant$deseq$counts %>% 
    rownames_to_column('gene') %>% 
    filter(grepl('RAS', gene))

```


### E: TE volcano
```{r}
te.de.volcano <- function(de.seq){
  de.seq <- 
    de.seq %>% 
    filter(!grepl('^MT', gene)) %>% 
    merge(te.families.clades, by = 'gene', all.x = T) %>% 
    mutate(clade = ifelse(gene %in% gencode.35.lnc.gene.names$gene, 'lncRNA', clade),
           family = ifelse(is.na(family), 'gencode', family),
           clade = ifelse(is.na(clade), 'gencode', clade))
  
  x.max <- round(max(abs(de.seq$log2FoldChange)))
  label.max <- 0.01
  print(head(de.seq %>% filter(clade %in% c('SINE', 'LTR', 'DNA', 'LINE'))))
  
  ggplot(de.seq %>% 
           filter(clade %in% c('SINE', 'LTR', 'DNA', 'LINE')), 
         aes(log2FoldChange, -log10(padj), color = clade)) +
      geom_point(alpha = 0.3, size = rel(1)) + 
      geom_rug(size = rel(1), alpha = 0.3) + 
      scale_x_continuous(limits = c(-x.max,x.max), expand = c(0,0.5),
                         breaks = c(-x.max/2, 0, x.max/2)) +
      scale_color_viridis_d(direction = -1) +
      xlab('log2(Fold Change)') +
      ylab('-log10(FDR)') + 
      geom_vline(xintercept = 0, linetype = 'dashed', alpha = 0.4) +
      geom_text_repel(aes(label = ifelse(padj < label.max & log2FoldChange <= -x.max/0.5| log2FoldChange >= x.max/0.5,
                                         as.character(gene), '')),
                              box.padding = unit('0.8', 'lines'),
                              segment.color = 'grey',
                              segment.alpha = 0.2,
                              color = 'black',
                              ylim = c(2, Inf), size = rel(2)) + 
    facet_wrap(~clade, nrow = 2, ncol = 2, shrink = T)
}
```
####
```{r}

te.de.volcano(ucsc.rmsk.salmon.quant$exoTIC_2D$deseq$de)
te.de.volcano(ucsc.rmsk.salmon.quant$exoTIC_3D$deseq$de %>% mutate(log2FoldChange = -log2FoldChange))
te.de.volcano(ucsc.rmsk.salmon.quant$rnaEasy_2D$deseq$de)
te.de.volcano(ucsc.rmsk.salmon.quant$rnaEasy_3D$deseq$de %>% mutate(log2FoldChange = -log2FoldChange))


te.de.volcano(ucsc.rmsk.salmon.quant$panc_v_covid$deseq$de)

ucsc.rmsk.salmon.quant$sars_v_mock$deseq$de %>% mutate(log2FoldChange = -log2FoldChange) %>% 
  merge(ucsc.rmsk.salmon.quant$covid_v_ctrl$deseq$de, by = 'gene', all = T) %>% 
    filter(!grepl('^MT', gene)) %>% 
  merge(te.families.clades, by = 'gene', all.x = T) %>% 
  mutate(clade = ifelse(gene %in% gencode.35.lnc.gene.names$gene, 'lncRNA', clade),
         family = ifelse(is.na(family), 'gencode', family),
         clade = ifelse(is.na(clade), 'gencode', clade)) %>% 
  filter(clade %in% c('SINE', 'LTR', 'DNA', 'LINE')) %>% 
  ggplot(aes(log2FoldChange.x, log2FoldChange.y, color = clade)) + 
  geom_point() +
  # geom_text_repel(aes(label = ifelse(abs(log2FoldChange.x) > 1 & abs(log2FoldChange.y) > 1, gene, ''))) +
  geom_text_repel(aes(label = gene)) + 
  scale_color_viridis_d(direction = -1) +
  geom_hline(yintercept = 0) + geom_vline(xintercept = 0) +
  facet_wrap(~clade, nrow = 2, ncol = 2, shrink = T)

ucsc.rmsk.salmon.quant$panc_v_ctrl$deseq$de %>% mutate(log2FoldChange = -log2FoldChange) %>% 
  merge(ucsc.rmsk.salmon.quant$kraskd_v_kras$deseq$de %>% mutate(log2FoldChange = -log2FoldChange), by = 'gene', all = T) %>% 
    filter(!grepl('^MT', gene)) %>% 
  merge(te.families.clades, by = 'gene', all.x = T) %>% 
  mutate(clade = ifelse(gene %in% gencode.35.lnc.gene.names$gene, 'lncRNA', clade),
         family = ifelse(is.na(family), 'gencode', family),
         clade = ifelse(is.na(clade), 'gencode', clade)) %>% 
  filter(clade %in% c('SINE', 'LTR', 'DNA', 'LINE')) %>% 
  ggplot(aes(log2FoldChange.x, log2FoldChange.y, color = clade)) + 
  geom_point() +
  # geom_text_repel(aes(label = ifelse(abs(log2FoldChange.x) > 1 & abs(-log2FoldChange.y) > 1, gene, ''))) + 
  geom_text_repel(aes(label = gene)) +
  scale_color_viridis_d(direction = -1) +
  geom_hline(yintercept = 0) + geom_vline(xintercept = 0) +
  facet_wrap(~clade, nrow = 2, ncol = 2, shrink = T)
    

```

### F: RNA editing
```{r}
rna.editing.summary <- read_csv(paste0(output.data.dir, 
                                       'rna.editing.out/alu.total.bed/summary_dir/EditingIndex.csv')) %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>%
  #separate(Sample, sep = '[.]', into = c('condition', 'rep')) %>% 
  select(Sample, noise = 'A2CEditingIndex',  signal = 'A2GEditingIndex') %>% 
  filter(!grepl('intra|exo', Sample)) %>% 
  mutate(condition = ifelse(grepl('panc', Sample), 'panc', 'ctrl'),
         Sample = sub('.sorted', '', Sample)) %>% 
  filter(Sample %in% tx.patient.biotype.sample.detection.count$sample)

rna.editing.summary <- read_csv(paste0(output.data.dir, 
                                       'rna.editing.out/alu.total.bed/summary_dir/EditingIndex.csv')) %>% 
  filter(StrandDecidingMethod == 'RefSeqThenMMSites') %>%
  filter(grepl('intra|exo', Sample)) %>% 
  separate(Sample, sep = '[.]', into = c('condition', 'rep', 'context', NA)) %>% 
  select(condition, rep, context, noise = 'A2CEditingIndex',  signal = 'A2GEditingIndex')
 

ggplot(rna.editing.summary,
       aes(reorder(context, signal), signal, color = condition)) + 
  geom_point() +
  geom_point(reorder(context, signal), noise, color = condition, shape = 'diamond', alpha = 0.4) + 
  xlab('') + ylab('A-to-I Editing') + 
  scale_x_discrete(expand = c(0,0.5)) + 
  theme(axis.text.x = element_text(hjust = 1, angle = 40))

ggplot(rna.editing.summary,
       aes(condition, signal, color = context)) + 
  geom_boxplot(position = position_dodge(width = 1)) +  
  stat_compare_means(method = "t.test", label.y = 1.7) +
  geom_sina(position = position_dodge(width = 1)) +
  xlab('ALU TOTAL') + ylab('A-to-I Editing') + 
  scale_x_discrete(expand = c(0,0.5)) + 
  scale_y_continuous(limits = c(0,2), breaks = c(0,0.25,0.3,0.35,0.4,0.45,0.5,1,1.5,2))
```

### Nanosight profiles
```{r}

exoTIC.3D.nanosight.df %>%
  ggplot(aes(size, y = mean, ymin = mean - sd, 
             ymax = mean + sd, group = condition, 
             fill = condition, color = condition)) + 
  geom_ribbon(alpha = 0.25, linetype = 'dotted') + 
  geom_line(size = (rel(1))) +
  geom_vline(xintercept = seq(0,350,50), linetype = 'dashed', alpha = 0.15) +
  scale_x_continuous(limits = c(0, 350), breaks = c(0,50,100,150,200,250,300,350)) +
  scale_color_manual('condition', values = c(viridis(3)[1], viridis(3)[2])) +
  scale_fill_manual('condition', values = c(viridis(3)[1], viridis(3)[2]), guide = F) + 
  stat_peaks(size = (rel(4))) + 
  stat_peaks(geom='text', vjust = -2, ignore_threshold = 0.45, size = rel(4)) +
  ylab('million particles per mL') +
  xlab('size (nm)') +
  # ylim(0, 80) +
  ggtitle('exoTIC 3D') +
  geom_segment(x = 30, xend = 150, y = 60, yend = 60, color = 'black') -> exoTIC.3D.nanosight.plt

exoTIC.2D.nanosight.df %>%
  ggplot(aes(size, y = mean, ymin = mean - sd, 
             ymax = mean + sd, group = condition, 
             fill = condition, color = condition)) + 
  geom_ribbon(alpha = 0.25, linetype = 'dotted') + 
  geom_line(size = (rel(1))) +
  geom_vline(xintercept = seq(0,350,50), linetype = 'dashed', alpha = 0.15) +
  scale_x_continuous(limits = c(0, 350), breaks = c(0,50,100,150,200,250,300,350)) +
  scale_color_manual('condition', values = c(viridis(3)[1], viridis(3)[2])) +
  scale_fill_manual('condition', values = c(viridis(3)[1], viridis(3)[2]), guide = F) + 
  stat_peaks(size = (rel(4))) +
  stat_peaks(geom='text', vjust = -2, ignore_threshold = 0.45, size = rel(4)) +
  ylab('million particles per mL') +
  xlab('size (nm)') +
  # ylim(0, 80) +
  ggtitle('exoTIC 2D') +
  geom_segment(x = 30, xend = 150, y = 80, yend = 80, color = 'black') -> exoTIC.2D.nanosight.plt



rnaEasy.amg.nanosight.df %>%
  ggplot(aes(size, y = value, group = dose, 
             fill = dose, color = dose)) + 
  geom_line(size = (rel(1))) +
  geom_vline(xintercept = seq(0,350,50), linetype = 'dashed', alpha = 0.15) +
  scale_x_continuous(limits = c(0, 350), breaks = c(0,50,100,150,200,250,300,350)) +
  scale_color_viridis_d() +
  scale_fill_viridis_d(guide = F) + 
  stat_peaks(size = (rel(4))) + 
  stat_peaks(geom='text', vjust = -2, ignore_threshold = 0.45, size = rel(4)) +
  ylab('million particles per mL') +
  xlab('size (nm)') +
  # ylim(0, 80) +
  geom_segment(x = 30, xend = 150, y = 850, yend = 850, color = 'black') + 
  facet_row(~context) -> rnaEasy.amg.nanosight.plt


rnaEasy.dmso.nanosight.df %>%
  ggplot(aes(size, y = mean, ymin = mean - sd, 
             ymax = mean + sd, group = context, 
             fill = context, color = context)) + 
  geom_ribbon(alpha = 0.25, linetype = 'dotted') + 
  geom_line(size = (rel(1))) +
  geom_vline(xintercept = seq(0,350,50), linetype = 'dashed', alpha = 0.15) +
  scale_x_continuous(limits = c(0, 350), breaks = c(0,50,100,150,200,250,300,350)) +
  scale_color_manual('condition', values = c(viridis(3)[1], viridis(3)[2])) +
  scale_fill_manual('condition', values = c(viridis(3)[1], viridis(3)[2]), guide = F) +
  stat_peaks(size = (rel(4))) + 
  stat_peaks(geom='text', vjust = -2, ignore_threshold = 0.45, size = rel(4)) +
  ylab('million particles per mL') +
  xlab('size (nm)') +
  geom_segment(x = 30, xend = 150, y = 1500, yend = 1500, color = 'black') -> rnaEasy.dmso.nanosight.plt

```
#### 
```{r}
exoTIC.3D.nanosight.plt + exoTIC.2D.nanosight.plt

rnaEasy.amg.nanosight.plt #+ rnaEasy.dmso.nanosight.plt

```



## Fig. 2
### rerwip PCA
```{r}
set.seed(234)
### Gencode PCA
gene.patient.normalized.counts.pca <-
  # patient.plasma.quant$gene$te.locus.gencode.34.salmon.out %>% 
  # dds.gxi.df %>%
  # dds.gxi.norm.counts %>% 
   ucsc.rmsk.salmon.quant$rnaEasy_all$deseq$gxi.vst.counts %>%
  # salmon.quant$covid_v_ctrl$deseq$counts %>%
  # vds.gxi.df %>% 
  # select(-gene, -biotype) %>% 
  # merge(gencode.35.tx.to.gene %>% select('ensg', 'gene') %>% distinct(), by = 'gene') %>%
  filter_at(vars(contains('AMG')), all_vars(. >= 5)) %>%
  rownames_to_column('gene') %>%
  # select(-ensg) #%>%
  # distinct() %>%
  # filter(gene %in% filter(yang.ts.score.gtex, tissue == 'Pancreas')$gene) %>% 
  # filter(! gene %in% gencode.35.tx.to.gene$gene) %>%
  # filter(gene %in% c(salmon.quant$covid_v_ctrl$deseq$de$gene)) %>%
  # filter(gene %in% c(ucsc.rmsk.salmon.quant$panc_v_ctrl$deseq$de$gene)) %>%
  # filter(grepl('^Alu|^LTR|^L1|^L2|^MER|^HERV', gene)) %>%
  # filter(gene %in% filter(te.families.clades, clade %in% c('SINE', "LINE", 'LTR'))$gene) %>%
  # merge(gencode.35.tx.to.gene %>% select(ensg, gene), by = 'ensg') %>%
  distinct() %>% 
  # mutate(gene = paste0(gene, '_', ensg)) %>%
  # select(gene, contains('.')) %>% 
  # filter(!grepl('^RP|^MT|)n', gene)) %>% #, ! gene %in% patient.plasma.quant$rfe.gender.profile$optVariables) %>%
  # filter(gene %in% filter(panc1.intra.quant$de.seq$te.locus.gencode.34.salmon.out, log2FoldChange <= -1)$gene) %>%
  mutate(gene.cv = rowSds(as.matrix(select(., contains('.'))))/rowMeans(select(., contains('.')))) %>% 
  filter(gene.cv >= median(gene.cv)) %>%
  select(-gene.cv) %>%
  column_to_rownames('gene') %>%
  # select(-ensg) %>% 
  t() %>% 
  as.data.frame() %>% 
  prcomp(scale=T, center = T)
  
summary(gene.patient.normalized.counts.pca)
pcs.of.interest <- apply(gene.patient.normalized.counts.pca$x, 2, var)
pcs.props <-  pcs.of.interest/sum(pcs.of.interest)
cumsum(pcs.props)[c(1,2)]

as.data.frame(pcs.props) %>% 
  rownames_to_column('pc') %>% 
  mutate(pc = as.numeric(sub('PC', '', pc))) %>% 
  ggplot(aes(pc, pcs.props)) + 
  geom_col()
 
# convert output to dataframe
pca.out <- as.data.frame(gene.patient.normalized.counts.pca$x) %>% 
  rownames_to_column('patient')

# summarize PC contributions
pca.out.summary <- 
  as.data.frame(summary(gene.patient.normalized.counts.pca)$importance) %>% 
  rownames_to_column('metric') %>% 
  gather(pc, value, -metric) %>% 
  spread(metric, value)
# summarize PC contributions
pca.out.contributions <- 
  as.data.frame(gene.patient.normalized.counts.pca$rotation) %>% 
  rownames_to_column('tx') %>% 
  gather(pc, contrib, -tx) %>% 
  group_by(pc) %>% 
  #merge(gencode.reference,
  #      by.x = 'tx',
  #      by.y = 'gene.name') %>% 
  group_by(pc) %>% 
  mutate(contrib = abs(contrib),
         rel.contrib = contrib/sum(contrib))

# plot PC1 and PC2
pca.out %>% 
  separate(patient, sep = '[.]', 
           into = c('platform', 'context', 'rep', 'condition')) %>%
  ggplot(aes(PC1, PC2,
             color = condition,
             shape = context)) + 
    geom_point(size = rel(4), alpha = 0.6) + 
    # scale_size_continuous(range = c(0.1, 2.5), breaks = c(0,5,10,15,20)) + 
    xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
    ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
    scale_color_manual(name = 'condition', values = viridis(3)[c(1,2)]) +
    scale_fill_aaas() +
    geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
    geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
    theme(legend.position = 'right',
            legend.key.width = unit(0.125,'cm'),
            legend.key.height = unit(0.4, 'cm'))


pca.out %>% 
  # rownames_to_column('sample') %>%
  # merge(current.meta.data, by = 'patient') %>% 
  # mutate(batch = as.factor(batch)) #%>% 
  # merge(merge(unsplice.df, splice.df, by = 'sample') %>% 
  #         mutate(unsplice_rate = unsplice_count / (unsplice_count + splice_count)),
  merge(unsplice.df,
        by.x = 'patient', by.y = 'sample') %>% 
  #separate(sample, sep = '[.]', into = c(NA, 'condition', NA)) %>% 
  ggplot(aes(PC1, PC2,
             color = unsplice_rate)) + 
    geom_point(size = rel(2)) + 
    scale_size_continuous(range = c(0.1, 2.5), breaks = c(0,5,10,15,20)) + 
    xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
    ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
    # scale_color_viridis_c(name = 'batch', alpha = 0.6) + 
    scale_color_gsea() +
    geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
    geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
    theme(legend.position = 'right',
            legend.key.width = unit(0.125,'cm'),
            legend.key.height = unit(0.4, 'cm')) 



# print top contibutors per PC
pca.out.contributions %>%
  group_by(pc) %>%
  filter(pc %in% c("PC1","PC2")) %>% 
  select(pc, rel.contrib, tx) %>% 
  # mutate(contrib = abs(contrib)) %>% 
  top_n(5, wt = rel.contrib) %>% 
  arrange(pc, -rel.contrib)
```


```{r}
library(uwot)

set.seed(234)

pca_umap <- as.data.frame(umap(pca.out, n_neighbors = 3, learning_rate = 0.5, init = "random"))

rownames(pca_umap) <- pca.out$patient

pca_umap %>% 
  rownames_to_column('sample') %>% 
  mutate(condition = sample) %>% 
  separate(condition, sep = '[.]', into = c('platform', 'context', NA, 'condition')) -> uwot.gencode.plot

ggplot(uwot.gencode.plot, 
       aes(V1, 
           V2, 
           color = condition, shape = context)) +
  geom_point() +
  ylab('U2') + 
  xlab('U1') +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.5, size = 0.25) + 
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.5, size = 0.25) + 
  scale_color_manual(name = 'condition', values = viridis(3)[c(1,2)]) +
  theme(legend.position = 'right',
        legend.spacing = unit(0, 'in'))

ggplot(uwot.gencode.plot, 
       aes(V1, 
           V2, 
           color = unsplice_rate,
           label = patient)) +
  geom_point() +
  ylab('U2') + 
  xlab('U1') +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.5, size = 0.25) + 
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.5, size = 0.25) + 
  # ggrepel::geom_text_repel(segment.color = NA,
  #                          point.padding = NA)  +
  # scale_color_manual(name = 'condition',
  #                    values = c('#37b578ff', '#440154ff')) +
  # scale_color_viridis_d() +
  scale_color_gsea() +
  theme(legend.position = 'right',
        legend.spacing = unit(0, 'in'))

```

#### rerwip Gencode Kmeans
```{r}
set.seed(234)

pca.for.clustering <- 
  pca.out %>% 
  select(patient, PC1:PC5) %>%
  column_to_rownames('patient')

# pca.for.clustering <- 
#   pca_umap


elbow.plt(pca.for.clustering)

kmeans(pca.for.clustering, 5, nstart = 50) -> pca.k.means

pca.for.clustering %>% 
  rownames_to_column('patient') %>% 
  merge(current.meta.data, by.x = 'patient', by.y = 'patient') %>% 
  mutate(cluster = factor(pca.k.means$cluster)) ->
  pca.for.clustering

pca.for.clustering %>% 
  ggplot(aes(PC1, 
             PC2, 
             color = cluster, 
             # shape = condition,
             label = patient)) +
             # label = paste(paste(patient, sep = '.'),
             #               paste0(diabetes, '-', gender, '-', age,'yrs'),
             #               sep = ', '))) + 
  geom_point(size = rel(2)) + 
  geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
  geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
  xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
  ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], 
                          digits = 3), sep = ' ')) +
  scale_color_viridis_d() #+ 
  # ggrepel::geom_text_repel(size = rel(3),
  #                          point.padding = 0.1,
  #                         box.padding = 0.2,
  #                         segment.color = NA,
  #                         color = 'black')

# panel.save(2, 'gender.filter.pca.kmeans.4')
```
### rerwip  NMF
```{r}
library(NMF)
set.seed(123)
salmon.tx.norm.for.nmf <- 
  ucsc.rmsk.salmon.quant$panc_v_ctrl$deseq$counts %>%
  rownames_to_column('gene') %>% 
  filter(gene %in% filter(ucsc.rmsk.salmon.quant$panc_v_ctrl$deseq$de, log2FoldChange <= -1)$gene) %>% 
  #filter_at(vars(contains('.')), all_vars(. >= 10)) %>% 
  # filter(gene %in% gen.34.lnc.names$gene, 
  #        gene %in% filter(patient.plasma.quant$de.seq$gencode.34.salmon.out, log2FoldChange > 1)$gene) %>%
  filter(!grepl('^RP|^HB|^MT|TMSB4X', gene)) %>%
  # filter(gene %in% filter(yang.ts.score.gtex, ! tissue %in% c('Cells', 'Whole Blood', 'Testis', 'Ovary', 'Uterus'))$gene) %>% 
  filter_at(vars(contains('.')), any_vars(. >= 1)) %>% 
  column_to_rownames('gene') %>% 
  t() %>% 
  as.data.frame()

# salmon.tx.norm.for.nmf <- 
#   patient.plasma.quant$gene$gencode.34.salmon.out %>% 
#   #filter_at(vars(contains('.')), all_vars(. >= 10)) %>% 
#   # filter(gene %in% gen.34.lnc.names$gene) %>% 
#   filter(!grepl('^HB|^MT|^S100|TMSB4X', gene)) %>% 
#   filter(gene %in% filter(yang.ts.score.gtex, ! tissue %in% c('Cells', 'Whole Blood', 'Testis', 'Ovary'))$gene) %>% 
#   filter_at(vars(contains('.')), any_vars(. >= 10)) %>% 
#   gather(sample, count, -gene) %>% 
#   merge(yang.ts.score.gtex, by = 'gene') %>% 
#   group_by(sample, tissue) %>% 
#   summarise(mean.counts = mean(count)) %>% 
#   spread(tissue, mean.counts) %>% 
#   column_to_rownames('sample') 

caret::nearZeroVar(salmon.tx.norm.for.nmf)

for(n in seq(2,5)){
  nmf.out <- nmf(salmon.tx.norm.for.nmf, n, nrun = 10)
  fit(nmf.out)
  print(paste(n, cophcor(nmf.out), sep = ':'))
}

nmf.out <- nmf(salmon.tx.norm.for.nmf, 5, nrun = 10)

as.data.frame(as.table(featureScore(nmf.out)[extractFeatures(nmf.out, 30L, 'list')[[1]]])) #%>% 
  merge(collison.subtype.genes,
        by.x = 'Var1',
        by.y = 'Genes') %>% 
  mutate(group = 'one') %>% 
  gather(type, score, -Var1, -group, -Freq) %>% 
  mutate(weighted.score = Freq * score) %>% 
  group_by(group, type) %>% 
  summarise(tot.score = mean(weighted.score) * 100)  -> group.1.nmf

as.data.frame(as.table(featureScore(nmf.out)[extractFeatures(nmf.out, 10L)[[2]]])) #%>% 
  merge(collison.subtype.genes,
        by.x = 'Var1',
        by.y = 'Genes') %>% 
  mutate(group = 'two') %>% 
  gather(type, score, -Var1, -group, -Freq) %>% 
  mutate(weighted.score = Freq * score) %>% 
  group_by(group, type) %>% 
  summarise(tot.score = mean(weighted.score) * 100) -> group.2.nmf

  
NMF::extractFeatures(nmf.out, method = 'max', nodups = T)

print(coef(nmf.out))

as.data.frame(coef(nmf.out)) %>% 
  rownames_to_column('basis') %>% 
  gather(gene, freq, -basis) %>% 
  group_by(gene) %>% 
  top_n(1, freq) %>% 
  group_by(basis) %>% 
  top_n(5, freq) %>% 
  

as.data.frame(coef(nmf.out)) %>% 
  rownames_to_column('basis') %>% 
  gather(gene, freq, -basis) %>% 
  merge(yang.ts.score.gtex, by = 'gene') %>% 
  group_by(basis, tissue) %>% 
  summarise(mean.freq = sum(freq)) %>% 
  group_by(basis) %>% 
  mutate(frac = mean.freq/sum(mean.freq)) %>% 
  select(-mean.freq) %>%
  ggplot(aes(reorder(tissue, -frac), frac, fill = basis)) + 
  geom_col(position = 'dodge') + theme(axis.text.x = element_text(angle = 40, hjust = 1))
  # spread(tissue, frac) %>% 
  # column_to_rownames('basis') %>% 
  # t() %>% as.data.frame() %>% rownames_to_column('tissue') 
  # mutate(diff = `1` - `2`) %>% 
  # ggplot(aes(reorder(tissue, -diff), diff)) + 
  # geom_col() + theme(axis.text.x = element_text(angle = 40, hjust = 1))

as.data.frame(coef(nmf.out)) %>% 
  rownames_to_column('basis') %>% 
  gather(gene, freq, -basis) %>% 
  merge(yang.ts.score.gtex, by = 'gene') %>% 
  filter(tissue == 'Spleen') %>% 
  arrange(-freq) %>% 
  select(gene, basis, freq) %>% 
  spread(basis, freq) %>% 
  mutate(diff = `1` - `2`) %>% 
  arrange(-diff)

as.data.frame(coef(nmf.out)) %>% 
  rownames_to_column('basis') %>% 
  gather(tissue, freq, -basis) %>% 
  # merge(yang.ts.score.gtex, by = 'gene') %>% 
  # filter(tissue == 'Spleen') %>% 
  arrange(-freq) %>% 
  # select(gene, basis, freq) %>% 
  ggplot(aes(reorder(tissue, -freq),freq, fill = basis)) + 
  geom_col(position = 'dodge') +
  theme(axis.text.x = element_text(angle = 40, hjust = 1))
  


qqnorm(featureScore(nmf.out))
qqline(featureScore(nmf.out))
basismap(nmf.out)
coefmap(nmf.out)
```
### F: TE Family dist
```{r}
te.only.anno.counts %>% 
  gather(sample, count, -gene, -family, -clade) %>% 
  separate(sample, sep = '[.]', into = c(NA, 'condition', NA)) %>% 
  group_by(condition, family) %>% 
  summarise(count = sum(count)) %>% 
  group_by(condition) %>% 
  mutate(freq = count / sum(count)) %>% 
  top_n(4, freq) -> te.family.dist.plt

ggplot(te.family.dist.plt,
       aes(family, freq, fill = condition, color = condition)) + 
  geom_bar(position = position_dodge(width = .95), stat = 'identity') +
  scale_fill_viridis_d(alpha = 0.6) +
  scale_color_viridis_d() + 
  scale_x_discrete(expand = c(0, 0.5)) +
  geom_hline(yintercept = seq(0,1,0.25), color = 'white') + 
  xlab('') + ylab('% of total te-aligned reads')

# panel.save(1, 'f.te.family.compare')
  
```  

### G: Tx length dist
```{r}
tx.normalized.counts %>% 
  gather(sample, count, -tx:-biotype) %>% 
  separate(sample, sep = '[.]', into = c(NA, 'condition', NA)) %>% 
  mutate(len = as.numeric(len)) %>% 
  group_by(tx, condition) %>% 
  summarize(count = sum(count), len = median(len)) %>% 
  filter(count > 0, 
         len <= 10000) %>% 
  ungroup() %>% 
  group_by(condition) %>% 
  mutate(freq = count / sum(count)) -> tx.len.dist.plt

ggplot(tx.len.dist.plt, 
        aes(len, log(count + 1), fill = condition, color = condition)) + 
  geom_col(size = rel(0.8)) +
  scale_x_continuous(expand = c(0, 0.5), n.breaks = 15) +
  scale_fill_viridis_d(alpha = 0.6) +
  scale_color_viridis_d() + 
  geom_hline(yintercept = seq(0,200,40), color = 'white') + 
  xlab('transcript length') + ylab('log(normalized counts + 1)')

panel.save(1, 'g.tx.len.dist', width = 'double')
```

### H: Mapping rates 
```{r}
patient.read.mapping %>% 
  filter(patient %in% colnames(patient.plasma.quant$gene$gencode.34.salmon.out)) %>% 
  gather(approach, rate, -patient) %>% 
  ggplot(aes(patient, rate, fill = approach)) + 
  geom_bar(stat = 'identity', position = 'dodge') +
  scale_fill_viridis_d(direction = -1) +
  scale_color_viridis_d(direction = -1) + 
  geom_hline(yintercept = seq(0,1,0.25), color = 'white') +
  scale_x_discrete(expand = c(0, 0.5)) +
  xlab('') + ylab('# of detected genes') + 
  theme(axis.text.x = element_text(angle = 40, hjust = 1))
panel.save('supplement', 'patient.mapping.rate.compare', width = 'double')

patient.star.unique.mapping %>% 
  # filter(patient %in% colnames(patient.plasma.quant$gene$gencode.34.salmon.out)) %>% 
  mutate(context = ifelse(grepl('intra', patient), 'intra', 
                          ifelse(grepl('exo', patient), 'exo',
                                 'plasma'))) %>% 
  ggplot(aes(reorder(patient, -unique.reads/1e6), unique.reads/1e6,fill = mapping_rate)) + 
  geom_bar(stat = 'identity', position = 'dodge') +
  scale_fill_viridis_c(direction = -1) +
  scale_color_viridis_c(direction = -1) + 
  #geom_hline(yintercept = seq(0,6,1), color = 'white') +
  scale_x_discrete(expand = c(0, 0.5)) +
  xlab('') + ylab('millions of reads (uniquely mapped)') + 
  theme(axis.text.x = element_text(angle = 40, hjust = 1)) + 
  facet_col(~context, scales = 'free')
panel.save('supplement', 'all.data.star.rate.and.amount', width = 'double', height = 'double')


patient.star.detailed.mapping %>% 
  mutate(condition = ifelse(grepl('panc', patient), 'panc', 'ctrl')) %>% 
  filter(patient %in% colnames(patient.plasma.quant$gene$gencode.34.salmon.out)) %>%
  ggplot(aes(reorder(condition, -multi_map_rate), multi_map_rate)) + 
  # geom_bar(stat = 'identity') +
  geom_sina() + 
  geom_boxplot(fill = NA) +
  scale_fill_viridis_c(direction = -1) +
  scale_color_viridis_c(direction = -1) + 
  #geom_hline(yintercept = seq(0,6,1), color = 'white') +
  scale_x_discrete(expand = c(0, 0.5)) +
  xlab('') + ylab('STAR multimapping rate')
panel.save('supplement', 'patient.multi.mapping.rate.dist', width = 'single', height = 'single')

```

## U18 HIV mapping
```{r}
hiv_mapping <- read_csv('scripts/eric.data/hiv_mapping.csv', col_names = c('sample', NA, 'rate')) %>% 
  separate(sample, sep = '_', into = c('sample', 'rep', NA))

ggplot(hiv_mapping, aes(reorder(sample, -rate), rate, color = sample, fill = sample)) + 
  geom_sina(alpha = 0.6, position = position_dodge(width = 1), size = rel(1.5)) +
  geom_boxplot(outlier.colour = NA, position = position_dodge(width = 1), fill = NA, size = rel(0.25)) +
  # geom_col() +
  # theme(axis.text.x = element_text(angle = 40, hjust = 1)) +
  scale_fill_d3() + 
  scale_color_d3() +
  # scale_fill_manual(values = c(viridis(1)[[1]], viridis(3)[[2]])) +
  # scale_color_manual(values = c(viridis(1)[[1]], viridis(3)[[2]])) +
  ylab('% unique mapping') + xlab('')

ggsave('scripts/eric.data/hiv_mapping.boxplot.pdf', width = 2.5, height = 2.5, units = 'in')
```

### COVID preliminary analysis
```{r}

znf.list <- read_csv('data/reference/all.human.znf.trono.db.csv') #,
  #                    col_names = c('gene',NA,NA,NA,NA,NA,NA,NA)) %>% 
  # select(gene)

covid.test.de <- 
  read_csv('scripts/eric.data/gencode.34.salmon.out.covid.covid.v.ctrl.de-seq.csv') %>% 
  rename(gene = 'X1') #%>% 
  # filter(padj < 0.05, abs(log2FoldChange) > 0.5)

gencode.de.volcano(covid.test.de)

ggsave('scripts/eric.data/covid.overall.gencode.volcano.pdf', width = 3, height = 3)

gencode.de.volcano(covid.test.de %>% filter(gene %in% gen.34.lnc.names$gene))

ggsave('scripts/eric.data/covid.lncRNA.gencode.volcano.pdf', width = 3, height = 3)

gencode.de.volcano(covid.test.de %>% filter(gene %in% znf.list$gene))

ggsave('scripts/eric.data/covid.znf.gencode.volcano.pdf', width = 3, height = 3)

gencode.de.volcano(covid.test.de %>% filter(gene %in% msig.df$HALLMARK_INFLAMMATORY_RESPONSE))

ggsave('scripts/eric.data/covid.hallmark.inflamm.gencode.volcano.pdf', width = 3, height = 3)

gencode.de.volcano(covid.test.de %>% filter(gene %in% msig.df$HALLMARK_KRAS_SIGNALING_UP))

ggsave('scripts/eric.data/covid.hallmark.kras.up.gencode.volcano.pdf', width = 3, height = 3)

gencode.de.volcano(covid.test.de %>% filter(gene %in% msig.df$REACTOME_MAPK_FAMILY_SIGNALING_CASCADES))

ggsave('scripts/eric.data/covid.hallmark.mapk.gencode.volcano.pdf', width = 3, height = 3)

# write.csv(file = 'scripts/eric.data/significant.gencode.te.de.csv', covid.test.te.de)


covid.test.te.de <- 
  read_csv('scripts/eric.data/te.locus.gencode.34.salmon.out.covid.covid.v.ctrl.de-seq.csv') %>% 
  rename(gene = 'X1') #%>% 
  # filter(padj < 0.05, abs(log2FoldChange) > 0.5)


# write.csv(file = 'scripts/eric.data/significant.gencode.de.csv', covid.test.de)

te.de.volcano(covid.test.te.de)
ggsave('scripts/eric.data/covid.te.clades.volcano.pdf', width = 3, height = 3)

covid.test.fgsea <- make.fgsea.ranks(covid.test.de)

covid.patient.metadata <- read_csv('/home/rreggiar/output.data/covid.patient.metadata.csv')

covid.patient.gene.norm.gencode.counts <- 
  read_csv('scripts/eric.data/gene.gencode.34.salmon.out.covid.covid.v.ctrl.de-seq.counts.csv') %>% 
  rename(gene = 'X1') %>% 
  gather(patient, norm.counts, -gene) %>% 
  filter(!grepl('ctrl', patient)) %>% 
  merge(covid.patient.metadata, by = 'patient')

covid.patient.gene.norm.gencode.counts %>% 
  filter(gene == 'BRAF') %>% 
  lm(norm.counts ~ tsd, .) %>% 
  summary()
  # filter(gene %in% covid.test.de$gene,
  #        gene %in% msig.df$HALLMARK_INFLAMMATORY_RESPONSE) %>% 
  # mutate(group = 'inflamm') %>% 
  # group_by(patient, tsd, age, group)
  ggplot(aes(tsd, log(norm.counts + 1))) + 
  geom_point() + 
  geom_line() + 
  stat_smooth(method = 'lm')

covid.patient.gene.norm.rmsk.gencode.counts <-
  read_csv('scripts/eric.data/gene.te.locus.gencode.34.salmon.out.covid.covid.v.ctrl.de-seq.counts.csv') %>% 
  rename(gene = 'X1') %>% 
  gather(patient, norm.counts, -gene) %>% 
  # filter(!grepl('ctrl', patient)) %>% 
  # merge(covid.patient.metadata, by = 'patient')
  mutate(condition = ifelse(grepl('covid', patient), 'covid', 'ctrl'))

ggplot(covid.patient.gene.norm.rmsk.gencode.counts %>% 
         filter(gene %in% c('CXCL8', 'IL7R', 'BRAF','ZNF91', 'HERVI-int', 'KRAS', 'L1HS', 'NFKB1','LTR10A')), 
       aes(reorder(gene, -norm.counts), log(norm.counts + 1), fill = condition, color = condition)) + 
  geom_jitter(alpha = 0.3, position = position_dodge(width = 1), size = rel(1.5)) + 
  geom_boxplot(outlier.colour = NA, position = position_dodge(width = 1), fill = NA, size = rel(0.25)) + 
  theme(axis.text.x = element_text(angle = 40, hjust = 1)) +
  scale_fill_manual(values = c(viridis(1)[[1]], viridis(3)[[2]])) + 
  scale_color_manual(values = c(viridis(1)[[1]], viridis(3)[[2]])) + 
  ylab('log(DESeqII norm. counts)') + xlab('')
ggsave('scripts/eric.data/covid.interesting.genes.dist.pdf', width = 3, height = 3)


covid.patient.gene.norm.rmsk.gencode.counts %>% 
  merge(te.families.clades, by = 'gene') %>% 
  filter(clade %in% c('SINE', 'LTR', 'DNA', 'LINE')) %>% 
  group_by(tsd, clade, patient, age) %>% 
  summarize(average.count = mean(norm.counts)) %>% 
  # filter(gene == 'HERVH-int') %>% 
  ggplot(aes(tsd, log(average.count + 1), group = clade, color = clade)) + 
  geom_point() + 
  geom_line() + 
  scale_color_viridis_d()

covid.patient.gene.norm.rmsk.gencode.counts %>% 
  distinct(patient, age, tsd) %>% 
  ggplot(aes(age, tsd)) +
  stat_smooth()

covid.patient.gene.norm.rmsk.gencode.counts %>% 
  filter(gene == 'HERVH-int') %>% 
  ggplot(aes(age, norm.counts)) + 
  geom_point() 
```
####
```{r}
gene.expression.hclust.plt <- function(gene.counts){
  gene.counts %>% 
    filter_at(vars(contains('covid.')), all_vars(. >= 10)) %>% 
    gather(patient, count, -gene) %>% 
    # merge(merged.patient.metadata %>% select(patient, batch), by.x = 'patient', by.y = 'patient') %>% 
    # filter(!grepl('^MT', gene)) %>% 
    mutate(count = log(count + 1),
           count = scale(count, center = T)) %>% 
    # unite('patient', c(patient, batch), sep = '_') %>% 
    spread(patient, count) %>% 
    column_to_rownames('gene') %>% 
    pheatmap(treeheight_row = 10, treeheight_col = 10, show_rownames = F, color = viridis(32), width = 5, height = 3)
}

covid.patient.gene.norm.gencode.counts <- 
  read_csv('scripts/eric.data/gene.gencode.34.salmon.out.covid.covid.v.ctrl.de-seq.counts.csv') %>% 
  rename(gene = 'X1')

covid.patient.gene.norm.rmsk.gencode.counts <-
  read_csv('scripts/eric.data/gene.te.locus.gencode.34.salmon.out.covid.covid.v.ctrl.de-seq.counts.csv') %>% 
  rename(gene = 'X1') 


pheat.1 <- gene.expression.hclust.plt(covid.patient.gene.norm.rmsk.gencode.counts)

save_pheatmap_pdf(x = pheat.1, 
                  filename = 'scripts/eric.data/covid.hclust.count.filter.10.pdf')


ggplot(covid.patient.gene.norm.rmsk.gencode.counts %>% 
         filter(gene %in% c('CXCL8', 'IL7R', 'BRAF',
                                                                                              'ZNF91', 'HERVH-int', 
                                                                                              'KRAS', 'IFNA', 'L1HS')), 
       aes(reorder(gene, -norm.counts), count, fill = type, color = type)) + 
  geom_sina(alpha = 0.3, position = position_dodge(width = 1), size = rel(1.5)) + 
  geom_boxplot(outlier.colour = NA, position = position_dodge(width = 1), fill = NA, size = rel(0.25)) + 
  # theme(axis.text.x = element_text(angle = 40, hjust = 1)) + 
  scale_fill_manual(values = c(viridis(1)[[1]], viridis(3)[[2]])) + 
  scale_color_manual(values = c(viridis(1)[[1]], viridis(3)[[2]])) + 
  ylab('DESeqII norm. counts') + xlab('')

```
####
```{r}
set.seed(234)
### Gencode PCA
gene.patient.normalized.counts.pca <-  
 covid.patient.gene.norm.rmsk.gencode.counts %>% 
  # select(-gene, -biotype) %>% 
  #filter_at(vars(contains('panc.')), all_vars(. >= 10)) %>% 
  filter(!grepl('^RP|^MT', gene), ! gene %in% patient.plasma.quant$rfe.gender.profile$optVariables) %>%
  # filter(gene %in% filter(panc1.intra.quant$de.seq$gencode.34.salmon.out, abs(log2FC) >= 3)$gene) %>%
  #   mutate(gene.cv = rowSds(as.matrix(.[, -c(1)]))/rowMeans(.[,-c(1)])) %>%
  mutate(gene.cv = rowSds(as.matrix(select(., contains('covid.'))))/rowMeans(select(., contains('covid.')))) %>%
  filter(gene.cv <= 2) %>%
  select(-gene.cv) %>%
  column_to_rownames('gene') %>%
  t() %>% 
  as.data.frame() %>% 
  prcomp(scale=T)

summary(gene.patient.normalized.counts.pca)
pcs.of.interest <- apply(gene.patient.normalized.counts.pca$x, 2, var)
pcs.props <-  pcs.of.interest/sum(pcs.of.interest)
cumsum(pcs.props)[1]

as.data.frame(pcs.props) %>% 
  rownames_to_column('pc') %>% 
  mutate(pc = as.numeric(sub('PC', '', pc))) %>% 
  ggplot(aes(pc, pcs.props)) + 
  geom_col()
 
# convert output to dataframe
pca.out <- as.data.frame(gene.patient.normalized.counts.pca$x) %>% 
  rownames_to_column('patient') %>% 
  merge(covid.patient.metadata, all.x = T) %>% 
  merge(patient.metadata, all.x = T, by = 'patient') %>% 
  mutate(gender.x = ifelse(is.na(gender.y), gender.x, gender.y),
         age.x = ifelse(is.na(age.y), age.x, age.y),
         condition = ifelse(is.na(condition), 'covid', condition))

# summarize PC contributions
pca.out.summary <- 
  as.data.frame(summary(gene.patient.normalized.counts.pca)$importance) %>% 
  rownames_to_column('metric') %>% 
  gather(pc, value, -metric) %>% 
  spread(metric, value)
# summarize PC contributions
pca.out.contributions <- 
  as.data.frame(gene.patient.normalized.counts.pca$rotation) %>% 
  rownames_to_column('tx') %>% 
  gather(pc, contrib, -tx) %>% 
  group_by(pc) %>% 
  #merge(gencode.reference,
  #      by.x = 'tx',
  #      by.y = 'gene.name') %>% 
  group_by(pc) %>% 
  mutate(contrib = abs(contrib)) %>% 
  filter(contrib > quantile(contrib, 0.95))

# plot PC1 and PC2
pca.out %>% 
  rownames_to_column('sample') %>% 
  #separate(sample, sep = '[.]', into = c(NA, 'condition', NA)) %>% 
  ggplot(aes(PC1, PC2,
             color = condition,
             label = tsd)) + 
    geom_point(size = rel(2)) + 
    scale_size_continuous(range = c(0.1, 2.5), breaks = c(0,5,10,15,20)) + 
    xlab(paste('PC1', round(cumsum(pcs.props)[1], digits = 3), sep = ' ')) +
    ylab(paste('PC2', round(cumsum(pcs.props)[2] - cumsum(pcs.props)[1], 
                            digits = 3), sep = ' ')) +
    # scale_color_viridis_d(name = 'condition', alpha = 0.6) +   
    scale_fill_manual(values = c(viridis(1)[[1]], viridis(3)[[2]])) + 
    scale_color_manual(values = c(viridis(1)[[1]], viridis(3)[[2]])) + 
    geom_vline(xintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) +
    geom_hline(yintercept = 0, linetype = 'dotted', alpha = 0.3, size = 0.25) + 
    theme(legend.position = 'right',
            legend.key.width = unit(0.125,'cm'),
            legend.key.height = unit(0.4, 'cm')) + 
  geom_text_repel(color = 'black', size = rel(0.5))
  
ggsave('scripts/eric.data/covid.pca.pdf', width = 3, height = 3)


# print top contibutors per PC
pca.out.contributions %>%
  group_by(pc) %>%
  filter(pc %in% c("PC1","PC2")) %>% 
  select(pc, contrib, tx) %>% 
  mutate(contrib = abs(contrib)) %>% 
  top_n(7, wt = abs(contrib)) %>% 
  arrange(pc, -contrib)
```
####
```{r}
covid.patient.gene.norm.gencode.counts <- 
  read_csv('scripts/eric.data/tx.te.locus.gencode.34.salmon.out.covid.covid.v.ctrl.de-seq.counts.csv') %>% 
  rename(gene = 'X1') %>% 
  separate(gene, sep = '\\|', into = c(NA,NA,NA,NA,'tx','gene','len','biotype')) #%>% 
  merge(gen.34.tx.2.gene, by = 'tx', all.x = T) %>% 
  merge(te.families.clades, by = 'tx', all.x = T) %>% 
  mutate(biotype = ifelse(is.na(biotype), clade, biotype))

  gen <- read_csv('scripts/eric.data/tx.te.locus.gencode.34.salmon.out.covid.covid.v.ctrl.de-seq.counts.csv') %>% 
    rename(gene = 'X1') %>% 
    filter_at(vars(contains('.')), any_vars(. >= 10)) %>% 
    filter(grepl('^ENST', gene)) %>% 
    separate(gene, sep = '\\|', into = c(NA,NA,NA,NA,'tx','gene','len','biotype'))
  rmsk <- read_csv('scripts/eric.data/tx.te.locus.gencode.34.salmon.out.covid.covid.v.ctrl.de-seq.counts.csv') %>% 
    rename(gene = 'X1') %>% 
    filter_at(vars(contains('.')), any_vars(. >= 10)) %>% 
    filter(!grepl('^ENST', gene)) %>% 
    separate(gene, sep = '_', into = c(NA, NA, 'gene', 'position', NA, NA, 'strand', NA)) %>% 
    mutate(position = sub('range=', '', position),
           strand = sub('strand=', '', strand)) %>% 
    merge(te.families.clades %>% distinct(), by = 'gene') %>% 
    mutate(tmp.gene = gene) %>% 
    unite('tx', c(gene, position, strand), sep = '_') %>% 
    select(tx, 'gene' = tmp.gene, 'len' = clade, 'biotype' = family, everything())
    
covid.tx <- rbind(gen, rmsk)

biotype.dist.plt(tx.counts = covid.tx, control = 'ctrl', treatment = 'covid')

ggsave('scripts/eric.data/covid.biotype.dist.pdf', width = 5, height = 3)

```
####
```{r}
trono.znf.targets <- read_csv('scripts/eric.data/all.znf.te.score.csv') %>% 
  gather(target, score, -X1) %>% 
  rename(gene = 'X1') %>% 
  filter(gene %in% filter(covid.test.de, log2FoldChange >= 1)$gene,
         score > 0) %>% 
  arrange(-score) %>% 
  separate(target, sep = '/', into = c('clade', 'family', 'te')) %>% 
  filter(te %in% filter(covid.test.te.de, log2FoldChange >= 1)$gene) %>% 
  merge(covid.test.te.de, by.x = 'te', by.y = 'gene') %>% 
  merge(covid.test.de, by.x = 'gene', by.y = 'gene') %>% 
  select(gene, te, clade, family, score, te.de='log2FoldChange.x', gene.de='log2FoldChange.y')


ggplot(trono.znf.targets %>% 
         group_by(gene) %>% 
         filter(score == max(score)), 
       aes(te.de, gene.de, label = paste(gene, te, clade, family, sep = '-'))) + 
  geom_point() + 
  geom_text_repel()
```



